{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis 08: GO Enrichment for Intervals"
   ],
   "id": "74affe57-c892-44ca-99b2-df9d67ef81ce"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": []
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Registered S3 methods overwritten by 'treeio':\n",
      "  method              from    \n",
      "  MRCA.phylo          tidytree\n",
      "  MRCA.treedata       tidytree\n",
      "  Nnode.treedata      tidytree\n",
      "  Ntip.treedata       tidytree\n",
      "  ancestor.phylo      tidytree\n",
      "  ancestor.treedata   tidytree\n",
      "  child.phylo         tidytree\n",
      "  child.treedata      tidytree\n",
      "  full_join.phylo     tidytree\n",
      "  full_join.treedata  tidytree\n",
      "  groupClade.phylo    tidytree\n",
      "  groupClade.treedata tidytree\n",
      "  groupOTU.phylo      tidytree\n",
      "  groupOTU.treedata   tidytree\n",
      "  is.rooted.treedata  tidytree\n",
      "  nodeid.phylo        tidytree\n",
      "  nodeid.treedata     tidytree\n",
      "  nodelab.phylo       tidytree\n",
      "  nodelab.treedata    tidytree\n",
      "  offspring.phylo     tidytree\n",
      "  offspring.treedata  tidytree\n",
      "  parent.phylo        tidytree\n",
      "  parent.treedata     tidytree\n",
      "  root.treedata       tidytree\n",
      "  rootnode.phylo      tidytree\n",
      "  sibling.phylo       tidytree"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "clusterProfiler v4.6.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'clusterProfiler'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.4     ✔ readr     2.1.5\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.1\n",
      "✔ ggplot2   3.5.1     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.3     ✔ tidyr     1.3.1\n",
      "✔ purrr     1.0.2     "
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::filter()   masks clusterProfiler::filter(), stats::filter()\n",
      "✖ dplyr::lag()      masks stats::lag()\n",
      "✖ purrr::simplify() masks clusterProfiler::simplify()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:lubridate':\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from 'package:glue':\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationDbi'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GOSemSim v2.24.0  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use GOSemSim in published research, please cite:\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu. Gene Ontology Semantic Similarity Analysis Using GOSemSim. In: Kidder B. (eds) Stem Cell Transcriptional Networks. Methods in Molecular Biology, 2020, 2117:207-215. Humana, New York, NY. doi:10.1007/978-1-0716-0301-7_11\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, Fei Li, Yide Qin, Xiaochen Bo, Yibo Wu, Shengqi Wang. GOSemSim: an R package for measuring semantic similarity among GO terms and gene products Bioinformatics 2010, 26(7):976-978. doi:10.1093/bioinformatics/btq064\n",
      "\n",
      "\n",
      "\n",
      "preparing gene to GO mapping data...\n",
      "\n",
      "preparing IC data..."
     ]
    }
   ],
   "source": [
    "library(clusterProfiler)\n"
   ],
   "id": "505d2d8e-07fb-403a-904e-74cfe27d2bb5"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_dir <- \"data/processed/20231116_Analysis_NemaScan\"\n",
    "tox_file <- \"data/processed/tox_data/tox_metadata.csv\"\n",
    "\n",
    "#### outputs ####\n",
    "out_dir <- \"data/processed/interval_genes/enrichments\"\n",
    "\n",
    "## create the output directory if it doesn't exist\n",
    "if (!dir.exists(out_dir)) {\n",
    "  dir.create(out_dir, recursive = TRUE)\n",
    "}\n"
   ],
   "id": "fccec483-ce3d-4876-a779-9a18269b7f8c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 9.33% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 2.07% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 16.34% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 11.49% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 10.74% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 17.57% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 8.37% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 8.75% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 24.08% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 7.55% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 8.54% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 7.46% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 3.28% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 5.88% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 18.29% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 6.58% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 1.73% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 11.22% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 4.92% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 4.78% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 5.23% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 0.63% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 9.66% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 15.79% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 18.02% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 11.58% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 4.96% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 18.28% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 15.56% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 6.26% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 6.13% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 23.71% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 16.37% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 8.3% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 3.47% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 7.65% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 16.67% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 11.46% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 7.26% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "'select()' returned 1:1 mapping between keys and columns"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in clusterProfiler::bitr(geneID = test_int_genes$WBGeneID, fromType =\n",
      "\"ENSEMBL\", : 20.15% of input gene IDs are fail to map..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in (function (..., deparse.level = 1) : number of columns of result is\n",
      "not a multiple of vector length (arg 3)"
     ]
    }
   ],
   "source": [
    "#### Main ####\n",
    "# load the toxicant metadata\n",
    "con_metadata <- data.table::fread(tox_file)\n",
    "\n",
    "# pull trait and nice drug label\n",
    "con_key <- con_metadata %>%\n",
    "  dplyr::select(\n",
    "    trait,\n",
    "    nice_drug_label2,\n",
    "    big_class,\n",
    "    moa_class\n",
    "  )\n",
    "\n",
    "# load the interval data\n",
    "inbred <- data.table::fread(\n",
    "  glue::glue(\"{ns_dir}/INBRED/Mapping/Processed/QTL_peaks_inbred.tsv\"),\n",
    "  data.table = F\n",
    ") %>%\n",
    "  # fix drug names\n",
    "  dplyr::mutate(\n",
    "    drug = stringr::str_replace(\n",
    "      trait,\n",
    "      pattern = \"^length_\",\n",
    "      replacement = \"\"\n",
    "    )\n",
    "  ) %>%\n",
    "  # remove the CV_ length_ traits\n",
    "  dplyr::filter(\n",
    "    !stringr::str_detect(drug, \"CV_\")\n",
    "  ) %>%\n",
    "  dplyr::select(\n",
    "    trait,\n",
    "    drug,\n",
    "    CHROM,\n",
    "    marker,\n",
    "    startPOS,\n",
    "    endPOS\n",
    "  ) %>%\n",
    "  # add the nice drug label\n",
    "  dplyr::left_join(con_key, by = c(\"trait\"))\n",
    "\n",
    "unique_traits <- unique(inbred$trait)\n",
    "\n",
    "# Initialize list to store results\n",
    "all_genes_dfs <- list()\n",
    "\n",
    "# Process each trait\n",
    "for (trait_id in unique_traits) {\n",
    "  # Get the nice drug label for this trait\n",
    "  nice_label <- inbred %>%\n",
    "    dplyr::filter(trait == trait_id) %>%\n",
    "    dplyr::pull(nice_drug_label2) %>%\n",
    "    unique()\n",
    "\n",
    "  # Get genes for each interval\n",
    "  genes_list <- get_genes_each_interval(\n",
    "    qtl = inbred,\n",
    "    trait_id = trait_id,\n",
    "    gff_df = gff,\n",
    "    cEGO = cEGO\n",
    "  )\n",
    "\n",
    "  # Convert to dataframe using nice drug label\n",
    "  genes_df <- genes_list_to_df(genes_list, nice_label)\n",
    "\n",
    "  # Store in results list\n",
    "  all_genes_dfs[[trait_id]] <- genes_df\n",
    "}\n"
   ],
   "id": "2683a9ce-7a3f-4dd4-b996-2647333cf58f"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
