{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis 13: Identify Candidate Genes"
   ],
   "id": "0abc8374-71a6-4b67-886e-4964a20c2782"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.4     ✔ readr     2.1.5\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.1\n",
      "✔ ggplot2   3.5.1     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.3     ✔ tidyr     1.3.1\n",
      "✔ purrr     1.0.2     \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::between()     masks data.table::between()\n",
      "✖ dplyr::filter()      masks stats::filter()\n",
      "✖ dplyr::first()       masks data.table::first()\n",
      "✖ lubridate::hour()    masks data.table::hour()\n",
      "✖ lubridate::isoweek() masks data.table::isoweek()\n",
      "✖ dplyr::lag()         masks stats::lag()\n",
      "✖ dplyr::last()        masks data.table::last()\n",
      "✖ lubridate::mday()    masks data.table::mday()\n",
      "✖ lubridate::minute()  masks data.table::minute()\n",
      "✖ lubridate::month()   masks data.table::month()\n",
      "✖ lubridate::quarter() masks data.table::quarter()\n",
      "✖ lubridate::second()  masks data.table::second()\n",
      "✖ purrr::transpose()   masks data.table::transpose()\n",
      "✖ lubridate::wday()    masks data.table::wday()\n",
      "✖ lubridate::week()    masks data.table::week()\n",
      "✖ lubridate::yday()    masks data.table::yday()\n",
      "✖ lubridate::year()    masks data.table::year()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    set_names\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    extract"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'kableExtra'\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    group_rows\n",
      "\n",
      "\n",
      "Attaching package: 'flextable'\n",
      "\n",
      "The following objects are masked from 'package:kableExtra':\n",
      "\n",
      "    as_image, footnote\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compose"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "library(tidyverse)\n"
   ],
   "id": "39dc6f8f-f658-4db4-b9db-0af7cc818e48"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs\n",
    "\n",
    "`ns_dir` - Path containing the NemaScan output directory. Used to load the BCSQ fine-mapping results `divergent_strain_bed` - Path to a divergent regions strain bed file"
   ],
   "id": "99bd8e64-40ed-42ea-a4cd-ce3a3d9a3171"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_dir <- \"data/processed/20231116_Analysis_NemaScan\"\n",
    "divergent_strain_bed <- \"data/raw/HDR/20240408_c_elegans_divergent_regions.bed\"\n"
   ],
   "id": "0c4e1aa2-bcc3-414a-8e3f-18ff0b026a58"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputs"
   ],
   "id": "869ab3b7-ab80-4326-b44d-efcda33537cd"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inbred_candidates_df_out_name <- \"data/processed/interval_genes/candidate_genes/inbred_candidate_genes.tsv.gz\"\n"
   ],
   "id": "c05174ef-7e29-491e-a75e-d2a280f72314"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ],
   "id": "b076ebb2-bdeb-43c2-8a66-f7d305f23dea"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### Functions ####\n",
    "\n",
    "joinDivergence <- function(pos.data, div.data, pos.strain) {\n",
    "  intersect <- valr::bed_intersect(\n",
    "    pos.data,\n",
    "    div.data\n",
    "  )\n",
    "  if (nrow(intersect) == 0) {\n",
    "    pos.data %>%\n",
    "      dplyr::mutate(\n",
    "        divergent = \"NOT_DIVERGENT\",\n",
    "        strain = pos.strain\n",
    "      )\n",
    "  } else {\n",
    "    intersect %>%\n",
    "      dplyr::mutate(divergent = \"DIVERGENT\") %>%\n",
    "      dplyr::select(chrom, start.x, end.x, divergent) %>%\n",
    "      dplyr::mutate(strain = pos.strain) %>%\n",
    "      dplyr::rename(\n",
    "        start = start.x,\n",
    "        end = end.x\n",
    "      )\n",
    "  }\n",
    "}\n",
    "\n",
    "# Function to get trait name from the file path\n",
    "get_traitname <- function(x) {\n",
    "  # get the name of the file\n",
    "  fn <- basename(x)\n",
    "\n",
    "  # remove the extension and file type info\n",
    "  id <- gsub(\"_bcsq_genes_inbred.tsv\", \"\", fn)\n",
    "  # print(id)\n",
    "\n",
    "\n",
    "  # Regular expression to match Roman numerals\n",
    "  chrom_pattern <- \"(?<=_)(I|II|III|IV|V|VI|X)(?=_)\"\n",
    "\n",
    "  # Split the string at the Roman numeral\n",
    "  parts <- strsplit(id, chrom_pattern, perl = TRUE)\n",
    "\n",
    "  # Extract the parts before and after the Roman numeral\n",
    "  trait <- sapply(parts, `[`, 1)\n",
    "  interval <- sapply(parts, `[`, 2)\n",
    "\n",
    "  # Remove the last underscore from 'before' and the first underscore from 'after'\n",
    "  trait <- sub(\"_$\", \"\", trait)\n",
    "  interval <- sub(\"^_\", \"\", interval)\n",
    "\n",
    "  # Extract the Roman numeral\n",
    "  chrom <- regmatches(x, regexpr(chrom_pattern, x, perl = TRUE))\n",
    "\n",
    "  # Return a list of before, after, and roman parts\n",
    "  list(trait = trait, interval = interval, chrom = chrom)\n",
    "}\n",
    "\n",
    "# # # test with just one value\n",
    "# values <- \"data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Paraquat_62_5_X_2674396-4285086\"\n",
    "# result <- get_traitname(values)\n",
    "# print(result$trait)\n",
    "# print(result$interval)\n",
    "# print(result$chrom)\n",
    "\n",
    "\n",
    "summarize.gene.candidates <- function(bcsq.file, LD, MAF, logPthresh, algorithm) {\n",
    "  # BCSQ data for a single QTL\n",
    "\n",
    "  # get the QTL information from the file path\n",
    "  trait_info <- get_traitname(bcsq.file)\n",
    "\n",
    "\n",
    "  bcsq.data <- data.table::fread(bcsq.file) %>%\n",
    "    dplyr::filter(\n",
    "      VARIANT_LD_WITH_PEAK_MARKER > LD,\n",
    "      MAF_variant > MAF,\n",
    "      VARIANT_IMPACT == \"HIGH\",\n",
    "      VARIANT_LOG10p > logPthresh\n",
    "    )\n",
    "\n",
    "  if (nrow(bcsq.data) == 0) {\n",
    "    print(\"no markers pass filter\")\n",
    "  } else {\n",
    "    # Nest by marker genotype - what is the breakdown of divergence within each genotype class?\n",
    "    nest.for.div <- bcsq.data %>%\n",
    "      dplyr::select(CHROM, POS, MARKER, STRAIN_GENOTYPE, STRAIN) %>%\n",
    "      dplyr::group_by(MARKER, STRAIN_GENOTYPE) %>%\n",
    "      tidyr::nest()\n",
    "\n",
    "    # x <- nest.for.div$data[[2]]\n",
    "    # nest.marker <- nest.for.div$MARKER[[2]]\n",
    "    # nest.genotype <- nest.for.div$STRAIN_GENOTYPE[[2]]\n",
    "    QTLdivergence <- function(x, nest.marker, nest.genotype) {\n",
    "      # lengthen for each strain and nest\n",
    "      marker.data <- data.frame(stringr::str_split(x$STRAIN, pattern = \",\")[[1]]) %>%\n",
    "        `colnames<-`(c(\"strain\")) %>%\n",
    "        cbind(x %>% dplyr::select(-STRAIN), .) %>%\n",
    "        dplyr::rename(chrom = CHROM) %>%\n",
    "        dplyr::mutate(\n",
    "          start = POS,\n",
    "          end = POS\n",
    "        ) %>%\n",
    "        dplyr::select(-POS) %>%\n",
    "        dplyr::filter(strain != \"N2\") %>%\n",
    "        dplyr::group_by(strain) %>%\n",
    "        tidyr::nest()\n",
    "\n",
    "      # pull divergent regions for each strain\n",
    "      div.test <- divergent_strain %>%\n",
    "        dplyr::filter(strain %in% unique(marker.data$strain)) %>%\n",
    "        dplyr::arrange(strain) %>%\n",
    "        dplyr::group_by(strain) %>%\n",
    "        tidyr::nest()\n",
    "\n",
    "      # # for testing\n",
    "      # out <- list()\n",
    "      # for(i in 1:length(marker.data$data)){\n",
    "      #   out[[i]] <- joinDivergence(pos.data = marker.data$data[[i]],\n",
    "      #                              div.data = div.test$data[[i]],\n",
    "      #                              pos.strain = marker.data$strain[[i]])\n",
    "      # }\n",
    "      # data.frame(Reduce(rbind, out)) %>%\n",
    "      #   dplyr::group_by(divergent) %>%\n",
    "      #   dplyr::count()\n",
    "\n",
    "      out <- purrr::pmap(\n",
    "        list(\n",
    "          marker.data$data,\n",
    "          div.test$data,\n",
    "          marker.data$strain\n",
    "        ),\n",
    "        joinDivergence\n",
    "      ) %>%\n",
    "        Reduce(rbind, .) %>%\n",
    "        dplyr::mutate(\n",
    "          MARKER = nest.marker,\n",
    "          ALLELE = nest.genotype\n",
    "        )\n",
    "      return(data.frame(out))\n",
    "    }\n",
    "    divergence.for.marker <- purrr::pmap(\n",
    "      .l = list(\n",
    "        nest.for.div$data,\n",
    "        nest.for.div$MARKER,\n",
    "        nest.for.div$STRAIN_GENOTYPE\n",
    "      ),\n",
    "      .f = QTLdivergence\n",
    "    ) %>%\n",
    "      Reduce(rbind, .) %>%\n",
    "      dplyr::group_by(MARKER, ALLELE, divergent) %>%\n",
    "      dplyr::count()\n",
    "\n",
    "    div.bcsq.data <- bcsq.data %>%\n",
    "      dplyr::distinct(MARKER, CHROM, POS, GENE_NAME, WBGeneID, PEAK_MARKER, VARIANT_LD_WITH_PEAK_MARKER, MAF_variant, VARIANT_IMPACT, VARIANT_LOG10p) %>%\n",
    "      dplyr::mutate(\n",
    "        trait = trait_info$trait,\n",
    "        interval = trait_info$interval,\n",
    "      ) %>%\n",
    "      # dplyr::arrange(-VARIANT_LOG10p, QTL) %>%\n",
    "      dplyr::left_join(., divergence.for.marker)\n",
    "    return(div.bcsq.data)\n",
    "  }\n",
    "}\n",
    "\n",
    "alleleDivergence <- function(data, marker, qtl) {\n",
    "  n.strains <- sum(data$n)\n",
    "  divergent.annotated <- data %>%\n",
    "    dplyr::mutate(\n",
    "      MARKER = marker,\n",
    "      QTL = qtl\n",
    "    ) %>%\n",
    "    dplyr::group_by(CHROM, POS, MAF_variant, GENE_NAME, WBGeneID, PEAK_MARKER, VARIANT_IMPACT, VARIANT_LD_WITH_PEAK_MARKER, VARIANT_LOG10p, MARKER, QTL, ALLELE) %>%\n",
    "    tidyr::pivot_wider(names_from = divergent, values_from = n)\n",
    "  divergent.annotated[is.na(divergent.annotated)] <- 0\n",
    "\n",
    "  if (\"DIVERGENT\" %in% colnames(divergent.annotated)) {\n",
    "    divergent.annotated %>%\n",
    "      dplyr::mutate(pct.allele.divergent = (DIVERGENT / (DIVERGENT + NOT_DIVERGENT)) * 100) %>%\n",
    "      dplyr::select(-DIVERGENT, -NOT_DIVERGENT) %>%\n",
    "      tidyr::pivot_wider(\n",
    "        id_cols = c(CHROM, POS, MAF_variant, GENE_NAME, WBGeneID, PEAK_MARKER, VARIANT_IMPACT, VARIANT_LD_WITH_PEAK_MARKER, VARIANT_LOG10p, QTL),\n",
    "        names_from = ALLELE,\n",
    "        values_from = pct.allele.divergent\n",
    "      ) %>%\n",
    "      dplyr::rename(\n",
    "        pct.divergent.ALT = ALT,\n",
    "        pct.divergent.REF = REF\n",
    "      )\n",
    "  } else {\n",
    "    divergent.annotated %>%\n",
    "      dplyr::ungroup() %>%\n",
    "      dplyr::distinct(CHROM, POS, MAF_variant, GENE_NAME, WBGeneID, PEAK_MARKER, VARIANT_IMPACT, VARIANT_LD_WITH_PEAK_MARKER, VARIANT_LOG10p, QTL) %>%\n",
    "      dplyr::mutate(\n",
    "        pct.divergent.ALT = 0,\n",
    "        pct.divergent.REF = 0\n",
    "      )\n",
    "  }\n",
    "}\n"
   ],
   "id": "c7a32038-5d5c-4d7a-998a-967a75f8c7e2"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main"
   ],
   "id": "f8e2af27-c27f-4893-803a-c547d56ca16d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Found 40 INBRED fine mapping BCSQ files"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Filtered to 40 INBRED length fine mapping BCSQ files"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\"\n",
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"no markers pass filter\"\n",
      "[1] \"no markers pass filter\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Joining with `by = join_by(MARKER)`\n",
      "Joining with `by = join_by(MARKER)`"
     ]
    }
   ],
   "source": [
    "# Pull the BCSQ files for LOCO and INBRED QTL fine mapping\n",
    "# This is confusing terms the LOCO files are loco QTL that have been fine mapped by inbred in NS pipeline\n",
    "# bcsq.loco <- list.files(\n",
    "#   path= glue::glue(\"{ns_dir}/LOCO/Fine_Mappings/Data\"),\n",
    "#   pattern = \"genes_loco\"\n",
    "#   )\n",
    "\n",
    "bcsq.inbred <- list.files(\n",
    "  path = glue::glue(\"{ns_dir}/INBRED/Fine_Mappings/Data\"),\n",
    "  pattern = \"genes_inbred\",\n",
    "  recursive = T,\n",
    "  full.names = T\n",
    ")\n",
    "\n",
    "n_bcsq.inbred <- length(bcsq.inbred)\n",
    "\n",
    "print(\n",
    "  glue::glue(\"Found {n_bcsq.inbred} INBRED fine mapping BCSQ files\")\n",
    ")\n"
   ],
   "id": "1139aff1-d4b9-4ef4-a960-c5ca823f5db7"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
