{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis 15: Candidate Gene Effects"
   ],
   "id": "2f2fa5f6-f700-439e-88e4-490f0cc78079"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'dplyr'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'data.table'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    between, first, last"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'kableExtra'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    group_rows"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'flextable'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:kableExtra':\n",
      "\n",
      "    as_image, footnote"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'rebus'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following object is masked from 'package:stringr':\n",
      "\n",
      "    regex"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'gtools'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following object is masked from 'package:rebus':\n",
      "\n",
      "    capture"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n"
   ],
   "id": "45ef02dc-93d8-4184-b4fd-08c5f4044a0c"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This script integrates the beta coefficient from the finemapping results with the candidate genes identifed in `candidate_gene_identification.qmd`.\n",
    "\n",
    "# Inputs\n",
    "\n",
    "File paths:"
   ],
   "id": "be0b96a0-b89a-4468-a852-6f4832d00397"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_table_fn <- \"data/processed/interval_genes/candidate_genes/inbred_candidate_genes.tsv.gz\"\n",
    "ns_folder <- \"data/processed/20231116_Analysis_NemaScan\"\n"
   ],
   "id": "8a00b494-1b72-497c-9578-da8a2baae76a"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Candidate gene table\n",
    "\n",
    "The candidate gene table generated in `candidate_gene_identification.qmd` contains information about SNVs located within candidate genes, along with allele and divergence information."
   ],
   "id": "8d0d810f-5f2b-4e2e-ace3-5be8abf1d9c5"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER\n",
      "        <char> <char>    <int>    <char>         <char>      <char>\n",
      "1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636\n",
      "2:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636\n",
      "3:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636\n",
      "4: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919\n",
      "5: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919\n",
      "6: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628\n",
      "   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p\n",
      "                         <num>       <num>         <char>          <num>\n",
      "1:                    0.776750   0.0666667           HIGH       7.687533\n",
      "2:                    0.776750   0.0666667           HIGH       7.687533\n",
      "3:                    0.776750   0.0666667           HIGH       7.687533\n",
      "4:                    1.000000   0.0512821           HIGH      13.064949\n",
      "5:                    1.000000   0.0512821           HIGH      13.064949\n",
      "6:                    0.797642   0.4871790           HIGH       5.487833\n",
      "             trait          interval ALLELE     divergent     n\n",
      "            <char>            <char> <char>        <char> <int>\n",
      "1:    length_2_4_D 13865698-15025162    ALT     DIVERGENT    11\n",
      "2:    length_2_4_D 13865698-15025162    ALT NOT_DIVERGENT     2\n",
      "3:    length_2_4_D 13865698-15025162    REF NOT_DIVERGENT   181\n",
      "4: length_Aldicarb   7790159-8989539    ALT NOT_DIVERGENT    10\n",
      "5: length_Aldicarb   7790159-8989539    REF NOT_DIVERGENT   184\n",
      "6: length_Aldicarb 16474675-17180844    ALT NOT_DIVERGENT    95"
     ]
    }
   ],
   "source": [
    "candidate_table <- data.table::fread(candidate_table_fn)\n",
    "head(candidate_table)\n"
   ],
   "id": "a6cf7194-0274-4d71-9ebd-2cc9850fdd3b"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The candidate table from candidate_gene_identification.qmd has duplicate rows for the same SNV (Single Nucleotide Variant at a specific genomic position). These duplicates exist because the table includes:\n",
    "\n",
    "-   ALLELE - the genotype (REF or ALT)\n",
    "-   divergent - divergence classification\n",
    "-   n - count of strains\n",
    "\n",
    "This means a single SNV in a specific gene appears multiple times - once for each combination of allele type and divergence status.\n",
    "\n",
    "Since this script is adding beta coefficients and allele information from the finemapping data (which already contains detailed allele info), it doesn’t need the genotype breakdown anymore"
   ],
   "id": "94153947-9eb8-49cf-9a84-4b1f3841d09d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] 410"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER\n",
      "        <char> <char>    <int>    <char>         <char>      <char>\n",
      "1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636\n",
      "2: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919\n",
      "3: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628\n",
      "4: IV_16738382     IV 16738382  Y51H4A.2 WBGene00013118  4:16631628\n",
      "5: IV_16867801     IV 16867801  T06A10.1 WBGene00235300  4:16631628\n",
      "6: IV_16989319     IV 16989319 Y116A8C.1 WBGene00013791  4:16631628\n",
      "   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p\n",
      "                         <num>       <num>         <char>          <num>\n",
      "1:                    0.776750   0.0666667           HIGH       7.687533\n",
      "2:                    1.000000   0.0512821           HIGH      13.064949\n",
      "3:                    0.797642   0.4871790           HIGH       5.487833\n",
      "4:                    1.000000   0.4564100           HIGH       7.687347\n",
      "5:                    0.940012   0.4717950           HIGH       6.174796\n",
      "6:                    0.940012   0.4717950           HIGH       6.174796\n",
      "             trait          interval\n",
      "            <char>            <char>\n",
      "1:    length_2_4_D 13865698-15025162\n",
      "2: length_Aldicarb   7790159-8989539\n",
      "3: length_Aldicarb 16474675-17180844\n",
      "4: length_Aldicarb 16474675-17180844\n",
      "5: length_Aldicarb 16474675-17180844\n",
      "6: length_Aldicarb 16474675-17180844"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] 179"
     ]
    }
   ],
   "source": [
    "n_row_before <- nrow(candidate_table)\n",
    "print(n_row_before)\n"
   ],
   "id": "dc7c49ce-d76b-4759-a3fa-053d072cb875"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finemapping data\n",
    "\n",
    "Load all of the finemapping data so that we can merge the beta coefficients into the candidate table.\n",
    "\n",
    "First define the functions to load the finemappings"
   ],
   "id": "638d4fa6-186f-497d-aac8-6c49d378aa06"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper functions to extract information from the finemap file names\n",
    "get_trait_value <- function(fn_name) {\n",
    "  # get trait value from file name (keeping the length_ prefix)\n",
    "  trait_value <- stringr::str_extract(fn_name, \"length_[^.]+\")\n",
    "  return(trait_value)\n",
    "}\n",
    "\n",
    "get_chromosome <- function(fn_name) {\n",
    "  # get chromosome from file name\n",
    "  chromosome <- stringr::str_extract(fn_name, \"\\\\b(I|II|III|IV|V|X)\\\\b\")\n",
    "  return(chromosome)\n",
    "}\n",
    "\n",
    "get_positions <- function(fn_name) {\n",
    "  # split the filename by period\n",
    "  parts <- stringr::str_split(fn_name, \"\\\\.\")[[1]]\n",
    "  startPOS <- as.numeric(parts[3])\n",
    "  stopPOS <- as.numeric(parts[4])\n",
    "  return(list(startPOS = startPOS, stopPOS = stopPOS))\n",
    "}\n",
    "\n",
    "# Main function\n",
    "aggregate_finemappings <- function(finemap_dir) {\n",
    "  finemap_files <- list.files(\n",
    "    path = finemap_dir,\n",
    "    pattern = \"finemap_inbred.(inbred|loco).fastGWA\",\n",
    "    full.names = TRUE,\n",
    "    recursive = TRUE\n",
    "  ) %>%\n",
    "    .[basename(.) %>% stringr::str_detect(\"^length\")]\n",
    "\n",
    "  finemap_df <- data.frame(\n",
    "    trait_value = character(),\n",
    "    CHROM = character(),\n",
    "    startPOS = numeric(),\n",
    "    endPOS = numeric(),\n",
    "    file_name = character(),\n",
    "    file_path = character(),\n",
    "    stringsAsFactors = FALSE\n",
    "  )\n",
    "\n",
    "  for (fn in finemap_files) {\n",
    "    fn_name <- basename(fn)\n",
    "    trait_value <- get_trait_value(fn_name)\n",
    "    chromosome <- get_chromosome(fn_name)\n",
    "    positions <- get_positions(fn_name)\n",
    "\n",
    "    finemap_df <- finemap_df %>%\n",
    "      add_row(\n",
    "        trait_value = trait_value,\n",
    "        CHROM = chromosome,\n",
    "        startPOS = positions$startPOS,\n",
    "        endPOS = positions$stopPOS,\n",
    "        file_name = fn_name,\n",
    "        file_path = fn\n",
    "      )\n",
    "  }\n",
    "\n",
    "  return(finemap_df)\n",
    "}\n"
   ],
   "id": "a04ae78f-a5ca-47fc-be26-75357765d96b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "      trait_value CHROM startPOS   endPOS\n",
      "1    length_2_4_D     V 13865698 15025162\n",
      "2 length_Aldicarb   III  7790159  8989539\n",
      "3 length_Aldicarb    IV 12966415 13439670\n",
      "4 length_Aldicarb    IV 16474675 17180844\n",
      "5 length_Aldicarb    IV  2179845  2507551\n",
      "6 length_Aldicarb     X 13002956 13505962\n",
      "                                                           file_name\n",
      "1     length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2  length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3 length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA\n",
      "4 length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "5   length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA\n",
      "6  length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA\n",
      "                                                                                                                               file_path\n",
      "1     data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA\n",
      "4 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "5   data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA\n",
      "6  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA"
     ]
    }
   ],
   "source": [
    "inbred_finemap_dir <-\n",
    "  glue::glue(\"{ns_folder}/INBRED/Fine_Mappings/Data\")\n",
    "\n",
    "inbred_finemap_df <- aggregate_finemappings(inbred_finemap_dir)\n",
    "head(inbred_finemap_df)\n"
   ],
   "id": "c2b45401-b0bf-42c5-a550-41c93b948ec0"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data frame contains the paths to all of the inbred finemapping results. Next we will filter to just the finemappings that correspond to the candidate genes in our candidate table and then read in those finemapping results into a single data frame."
   ],
   "id": "55b72667-3b72-4752-8c75-bcf79bb7a3d8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "      trait_value CHROM startPOS   endPOS\n",
      "1    length_2_4_D     V 13865698 15025162\n",
      "2 length_Aldicarb   III  7790159  8989539\n",
      "3 length_Aldicarb    IV 12966415 13439670\n",
      "4 length_Aldicarb    IV 16474675 17180844\n",
      "5 length_Aldicarb    IV  2179845  2507551\n",
      "6 length_Aldicarb     X 13002956 13505962\n",
      "                                                           file_name\n",
      "1     length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2  length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3 length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA\n",
      "4 length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "5   length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA\n",
      "6  length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA\n",
      "                                                                                                                               file_path\n",
      "1     data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA\n",
      "4 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "5   data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA\n",
      "6  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA\n",
      "           interval\n",
      "1 13865698-15025162\n",
      "2   7790159-8989539\n",
      "3 12966415-13439670\n",
      "4 16474675-17180844\n",
      "5   2179845-2507551\n",
      "6 13002956-13505962"
     ]
    }
   ],
   "source": [
    "# Create a column in the finemap df that matches the interval in the candidate table\n",
    "inbred_finemap_df_m <- inbred_finemap_df %>%\n",
    "  dplyr::mutate(interval = paste0(startPOS, \"-\", endPOS))\n",
    "head(inbred_finemap_df_m)\n"
   ],
   "id": "c8a0b27f-de65-46d5-9ff9-1659de35135b"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a table of unique trait and interval combinations from the candidate table"
   ],
   "id": "387e5b07-43ee-4721-a103-34cc7378087d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "                     trait          interval\n",
      "                    <char>            <char>\n",
      "1:            length_2_4_D 13865698-15025162\n",
      "2:         length_Aldicarb   7790159-8989539\n",
      "3:         length_Aldicarb 16474675-17180844\n",
      "4:         length_Aldicarb 13002956-13505962\n",
      "5: length_Arsenic_trioxide 13521931-14347284\n",
      "6: length_Arsenic_trioxide     28947-2527992"
     ]
    }
   ],
   "source": [
    "candidate_intervals <- candidate_table %>%\n",
    "  dplyr::select(trait, interval) %>%\n",
    "  dplyr::distinct()\n",
    "head(candidate_intervals)\n"
   ],
   "id": "d7d1837a-0f6f-43da-90f7-9afb1e67e945"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the finemapping data frame to only include files that match the candidate intervals"
   ],
   "id": "4f248ad9-4a5a-449b-bb11-739339bbaeda"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "              trait_value CHROM startPOS   endPOS\n",
      "1            length_2_4_D     V 13865698 15025162\n",
      "2         length_Aldicarb   III  7790159  8989539\n",
      "3         length_Aldicarb    IV 16474675 17180844\n",
      "4         length_Aldicarb     X 13002956 13505962\n",
      "5 length_Arsenic_trioxide    II 13521931 14347284\n",
      "6 length_Arsenic_trioxide   III    28947  2527992\n",
      "                                                                   file_name\n",
      "1             length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2          length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3         length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "4          length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA\n",
      "5 length_Arsenic_trioxide.II.13521931.14347284.finemap_inbred.inbred.fastGWA\n",
      "6    length_Arsenic_trioxide.III.28947.2527992.finemap_inbred.inbred.fastGWA\n",
      "                                                                                                                                       file_path\n",
      "1             data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA\n",
      "2          data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA\n",
      "3         data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA\n",
      "4          data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA\n",
      "5 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Arsenic_trioxide.II.13521931.14347284.finemap_inbred.inbred.fastGWA\n",
      "6    data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Arsenic_trioxide.III.28947.2527992.finemap_inbred.inbred.fastGWA\n",
      "           interval\n",
      "1 13865698-15025162\n",
      "2   7790159-8989539\n",
      "3 16474675-17180844\n",
      "4 13002956-13505962\n",
      "5 13521931-14347284\n",
      "6     28947-2527992"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] 31"
     ]
    }
   ],
   "source": [
    "filtered_finemap_df <- inbred_finemap_df_m %>%\n",
    "  dplyr::inner_join(candidate_intervals, by = c(\"trait_value\" = \"trait\", \"interval\" = \"interval\"))\n",
    "head(filtered_finemap_df)\n"
   ],
   "id": "4c0f10e6-09f1-423d-a5bb-5207eaab60e4"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in all of the filtered finemapping results and combine them into a single data frame"
   ],
   "id": "c9d4f2d0-b9aa-4de0-a5eb-4c6f423c3797"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "# A tibble: 6 × 18\n",
      "  trait_value  CHROM startPOS   endPOS file_name  file_path interval   CHR SNP  \n",
      "  <chr>        <chr>    <dbl>    <dbl> <chr>      <chr>     <chr>    <int> <chr>\n",
      "1 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "2 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "3 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "4 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "5 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "6 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…\n",
      "# ℹ 9 more variables: POS <int>, A1 <chr>, A2 <chr>, N <int>, AF1 <dbl>,\n",
      "#   BETA <dbl>, SE <dbl>, P <dbl>, trait <chr>"
     ]
    }
   ],
   "source": [
    "combine_finemapping_files <- function(file_df) {\n",
    "  combined_df <- file_df %>%\n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "      data = list(data.table::fread(file_path) %>%\n",
    "        mutate(trait = trait_value))\n",
    "    ) %>%\n",
    "    tidyr::unnest(cols = c(data))\n",
    "\n",
    "  return(combined_df)\n",
    "}\n",
    "\n",
    "filtered_inbred_finemap_data <- combine_finemapping_files(filtered_finemap_df)\n",
    "head(filtered_inbred_finemap_data)\n"
   ],
   "id": "a4d77039-762d-40cf-82d9-8a7ffa4117b5"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now format and pull the relevant columns from the combined finemapping data. Pulling the SNP genotypes (A1, A2) from the finemapping output. Per [GCTA inbred output docs](https://yanglab.westlake.edu.cn/software/gcta/#fastGWA)"
   ],
   "id": "1f5606ff-e3bb-4fe2-99e9-2e287105143c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "# A tibble: 6 × 5\n",
      "  trait        MARKER     beta_coefficient effect_allele other_allele\n",
      "  <chr>        <chr>                 <dbl> <chr>         <chr>       \n",
      "1 length_2_4_D V_13865698            0.493 G             T           \n",
      "2 length_2_4_D V_13865823            5.33  A             C           \n",
      "3 length_2_4_D V_13865980            2.87  A             G           \n",
      "4 length_2_4_D V_13867914            3.67  C             T           \n",
      "5 length_2_4_D V_13868694           -3.89  T             C           \n",
      "6 length_2_4_D V_13869222            8.62  G             A           "
     ]
    }
   ],
   "source": [
    "finemap_snvs <- filtered_inbred_finemap_data %>%\n",
    "  dplyr::select(trait, SNP, BETA, A1, A2) %>%\n",
    "  dplyr::rename(\n",
    "    MARKER = SNP,\n",
    "    beta_coefficient = BETA,\n",
    "    effect_allele = A1,\n",
    "    other_allele = A2\n",
    "  ) %>%\n",
    "  dplyr::mutate(MARKER = sapply(MARKER, convert_marker, to_marker = \"RM\", from_marker = \"NUM\")) %>%\n",
    "  dplyr::mutate(MARKER = stringr::str_replace_all(MARKER, \":\", \"_\"))\n",
    "\n",
    "head(finemap_snvs)\n"
   ],
   "id": "7d0277e6-0ab3-458e-a857-f3edd2cf1190"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main\n",
    "\n",
    "Join the candidate table with the finemapping SNV data to add the beta coefficients and allele information"
   ],
   "id": "f22eaace-1c15-459d-b3b3-c223098fece6"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER\n",
      "        <char> <char>    <int>    <char>         <char>      <char>\n",
      "1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636\n",
      "2: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919\n",
      "3: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628\n",
      "4: IV_16738382     IV 16738382  Y51H4A.2 WBGene00013118  4:16631628\n",
      "5: IV_16867801     IV 16867801  T06A10.1 WBGene00235300  4:16631628\n",
      "6: IV_16989319     IV 16989319 Y116A8C.1 WBGene00013791  4:16631628\n",
      "   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p\n",
      "                         <num>       <num>         <char>          <num>\n",
      "1:                    0.776750   0.0666667           HIGH       7.687533\n",
      "2:                    1.000000   0.0512821           HIGH      13.064949\n",
      "3:                    0.797642   0.4871790           HIGH       5.487833\n",
      "4:                    1.000000   0.4564100           HIGH       7.687347\n",
      "5:                    0.940012   0.4717950           HIGH       6.174796\n",
      "6:                    0.940012   0.4717950           HIGH       6.174796\n",
      "             trait          interval beta_coefficient effect_allele\n",
      "            <char>            <char>            <num>        <char>\n",
      "1:    length_2_4_D 13865698-15025162          16.5498             A\n",
      "2: length_Aldicarb   7790159-8989539         -54.6711             T\n",
      "3: length_Aldicarb 16474675-17180844          16.5350             A\n",
      "4: length_Aldicarb 16474675-17180844         -20.9067             G\n",
      "5: length_Aldicarb 16474675-17180844         -18.4316             T\n",
      "6: length_Aldicarb 16474675-17180844         -18.4316             G\n",
      "   other_allele\n",
      "         <char>\n",
      "1:            T\n",
      "2:            C\n",
      "3:            T\n",
      "4:            T\n",
      "5:            G\n",
      "6:            A"
     ]
    }
   ],
   "source": [
    "candidate_table_with_betas <- candidate_table %>%\n",
    "  dplyr::left_join(finemap_snvs, by = c(\"trait\", \"MARKER\"))\n",
    "head(candidate_table_with_betas)\n"
   ],
   "id": "cfb41770-f721-48ba-8039-9bb87e727981"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputs\n",
    "\n",
    "Save the candidate table with beta coefficients"
   ],
   "id": "cadca480-92b3-4e53-9dcd-1c5330f49be2"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_table_betas_outfn <- \"data/processed/interval_genes/candidate_genes/inbred_candidate_genes_with_betas.tsv.gz\"\n",
    "save_compressed_tsv(candidate_table_with_betas, candidate_table_betas_outfn)\n"
   ],
   "id": "6815d45b-a9fb-4515-b3e0-76bb3f76a11d"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
