<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
    <meta charset="utf-8">
    <meta name="generator" content="quarto-1.8.23">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


    <title>Analysis 15: Candidate Gene Effects</title>
    <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      div.columns{display: flex; gap: min(4vw, 1.5em);}
      div.column{flex: auto; overflow-x: auto;}
      div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
      ul.task-list{list-style: none;}
      ul.task-list li input[type="checkbox"] {
        width: 0.8em;
        margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
        vertical-align: middle;
      }
      /* CSS for syntax highlighting */
      html { -webkit-text-size-adjust: 100%; }
      pre > code.sourceCode { white-space: pre; position: relative; }
      pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
      pre > code.sourceCode > span:empty { height: 1.2em; }
      .sourceCode { overflow: visible; }
      code.sourceCode > span { color: inherit; text-decoration: inherit; }
      div.sourceCode { margin: 1em 0; }
      pre.sourceCode { margin: 0; }
      @media screen {
      div.sourceCode { overflow: auto; }
      }
      @media print {
      pre > code.sourceCode { white-space: pre-wrap; }
      pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
      }
      pre.numberSource code
        { counter-reset: source-line 0; }
      pre.numberSource code > span
        { position: relative; left: -4em; counter-increment: source-line; }
      pre.numberSource code > span > a:first-child::before
        { content: counter(source-line);
          position: relative; left: -1em; text-align: right; vertical-align: baseline;
          border: none; display: inline-block;
          -webkit-touch-callout: none; -webkit-user-select: none;
          -khtml-user-select: none; -moz-user-select: none;
          -ms-user-select: none; user-select: none;
          padding: 0 4px; width: 4em;
        }
      pre.numberSource { margin-left: 3em;  padding-left: 4px; }
      div.sourceCode
        {   }
      @media screen {
      pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
      }
    </style>

    <style>
      body.hypothesis-enabled #quarto-embed-header {
        padding-right: 36px;
      }

      #quarto-embed-header {
        height: 3em;
        width: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        border-bottom: solid 1px;
      }

      #quarto-embed-header h6 {
        font-size: 1.1em;
        padding-top: 0.6em;
        margin-left: 1em;
        margin-right: 1em;
        font-weight: 400;
      }

      #quarto-embed-header a.quarto-back-link,
      #quarto-embed-header a.quarto-download-embed {
        font-size: 0.8em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 1em;
        margin-right: 1em;
      }

      .quarto-back-container {
        padding-left: 0.5em;
        display: flex;
      }

      .headroom {
          will-change: transform;
          transition: transform 200ms linear;
      }

      .headroom--pinned {
          transform: translateY(0%);
      }

      .headroom--unpinned {
          transform: translateY(-100%);
      }      
    </style>

    <script>
    window.document.addEventListener("DOMContentLoaded", function () {

      var header = window.document.querySelector("#quarto-embed-header");
      const titleBannerEl = window.document.querySelector("body > #title-block-header");
      if (titleBannerEl) {
        titleBannerEl.style.paddingTop = header.clientHeight + "px";
      }
      const contentEl = window.document.getElementById('quarto-content');
      for (const child of contentEl.children) {
        child.style.paddingTop = header.clientHeight + "px";
        child.style.marginTop = "1em";
      }

      // Use the article root if the `back` call doesn't work. This isn't perfect
      // but should typically work
      window.quartoBackToArticle = () => {
        var currentUrl = window.location.href;
        window.history.back();
        setTimeout(() => {
            // if location was not changed in 100 ms, then there is no history back
            if(currentUrl === window.location.href){              
                // redirect to site root
                window.location.href = "../index.html";
            }
        }, 100);
      }

      const headroom = new window.Headroom(header, {
        tolerance: 5,
        onPin: function () {
        },
        onUnpin: function () {
        },
      });
      headroom.init();
    });
    </script>

    
<script src="../site_libs/manuscript-notebook/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-1fe81d0376b2c50856e68e651e390326.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a74871fe4945b66d259aafc266475145.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
     
      </head>

  <body class="quarto-notebook quarto-light">
    <div id="quarto-embed-header" class="headroom fixed-top bg-primary">
      
      <a onclick="window.quartoBackToArticle(); return false;" class="btn btn-primary quarto-back-link" href=""><i class="bi bi-caret-left"></i> Back to Article</a>
      <h6><i class="bi bi-journal-code"></i> Analysis 15: Candidate Gene Effects</h6>

            <a href="../analysis/candidate_gene_effects.qmd" class="btn btn-primary quarto-download-embed" download="candidate_gene_effects.qmd">Download Source</a>
          </div>

     <header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Analysis 15: Candidate Gene Effects</h1>
          </div>

    
    <div class="quarto-title-meta-container">
      <div class="quarto-title-meta-column-start">
        
        <div class="quarto-title-meta">

                
          
                
              </div>
      </div>
      <div class="quarto-title-meta-column-end quarto-other-formats-target">
      </div>
    </div>



    <div class="quarto-other-links-text-target">
    </div>  </div>
</header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#inputs" id="toc-inputs" class="nav-link" data-scroll-target="#inputs">Inputs</a>
  <ul class="collapse">
  <li><a href="#candidate-gene-table" id="toc-candidate-gene-table" class="nav-link" data-scroll-target="#candidate-gene-table">Candidate gene table</a></li>
  <li><a href="#finemapping-data" id="toc-finemapping-data" class="nav-link" data-scroll-target="#finemapping-data">Finemapping data</a></li>
  </ul></li>
  <li><a href="#main" id="toc-main" class="nav-link" data-scroll-target="#main">Main</a></li>
  <li><a href="#outputs" id="toc-outputs" class="nav-link" data-scroll-target="#outputs">Outputs</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">      

       <div class="cell-container"><div class="cell-decorator"><pre>In [1]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'data.table'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    between, first, last</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"bin/outs.R"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'kableExtra'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    group_rows</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'flextable'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:kableExtra':

    as_image, footnote</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"bin/general.R"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'rebus'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stringr':

    regex</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'gtools'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:rebus':

    capture</code></pre>
</div>
</div></div>
<section id="overview" class="level1">
<h1>Overview</h1>
<p>This script integrates the beta coefficient from the finemapping results with the candidate genes identifed in <code>candidate_gene_identification.qmd</code>.</p>
</section>
<section id="inputs" class="level1">
<h1>Inputs</h1>
<p>File paths:</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [2]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>candidate_table_fn <span class="ot">&lt;-</span> <span class="st">"data/processed/interval_genes/candidate_genes/inbred_candidate_genes.tsv.gz"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>ns_folder <span class="ot">&lt;-</span> <span class="st">"data/processed/20231116_Analysis_NemaScan"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
<section id="candidate-gene-table" class="level2">
<h2 class="anchored" data-anchor-id="candidate-gene-table">Candidate gene table</h2>
<p>The candidate gene table generated in <code>candidate_gene_identification.qmd</code> contains information about SNVs located within candidate genes, along with allele and divergence information.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [3]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>candidate_table <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(candidate_table_fn)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(candidate_table)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER
        &lt;char&gt; &lt;char&gt;    &lt;int&gt;    &lt;char&gt;         &lt;char&gt;      &lt;char&gt;
1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636
2:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636
3:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636
4: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919
5: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919
6: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628
   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p
                         &lt;num&gt;       &lt;num&gt;         &lt;char&gt;          &lt;num&gt;
1:                    0.776750   0.0666667           HIGH       7.687533
2:                    0.776750   0.0666667           HIGH       7.687533
3:                    0.776750   0.0666667           HIGH       7.687533
4:                    1.000000   0.0512821           HIGH      13.064949
5:                    1.000000   0.0512821           HIGH      13.064949
6:                    0.797642   0.4871790           HIGH       5.487833
             trait          interval ALLELE     divergent     n
            &lt;char&gt;            &lt;char&gt; &lt;char&gt;        &lt;char&gt; &lt;int&gt;
1:    length_2_4_D 13865698-15025162    ALT     DIVERGENT    11
2:    length_2_4_D 13865698-15025162    ALT NOT_DIVERGENT     2
3:    length_2_4_D 13865698-15025162    REF NOT_DIVERGENT   181
4: length_Aldicarb   7790159-8989539    ALT NOT_DIVERGENT    10
5: length_Aldicarb   7790159-8989539    REF NOT_DIVERGENT   184
6: length_Aldicarb 16474675-17180844    ALT NOT_DIVERGENT    95</code></pre>
</div>
</div></div>
<p>The candidate table from candidate_gene_identification.qmd has duplicate rows for the same SNV (Single Nucleotide Variant at a specific genomic position). These duplicates exist because the table includes:</p>
<ul>
<li>ALLELE - the genotype (REF or ALT)</li>
<li>divergent - divergence classification</li>
<li>n - count of strains</li>
</ul>
<p>This means a single SNV in a specific gene appears multiple times - once for each combination of allele type and divergence status.</p>
<p>Since this script is adding beta coefficients and allele information from the finemapping data (which already contains detailed allele info), it doesn’t need the genotype breakdown anymore</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [4]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>n_row_before <span class="ot">&lt;-</span> <span class="fu">nrow</span>(candidate_table)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(n_row_before)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 410</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>candidate_table <span class="ot">&lt;-</span> candidate_table <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(ALLELE, divergent, n)) <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(candidate_table)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER
        &lt;char&gt; &lt;char&gt;    &lt;int&gt;    &lt;char&gt;         &lt;char&gt;      &lt;char&gt;
1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636
2: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919
3: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628
4: IV_16738382     IV 16738382  Y51H4A.2 WBGene00013118  4:16631628
5: IV_16867801     IV 16867801  T06A10.1 WBGene00235300  4:16631628
6: IV_16989319     IV 16989319 Y116A8C.1 WBGene00013791  4:16631628
   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p
                         &lt;num&gt;       &lt;num&gt;         &lt;char&gt;          &lt;num&gt;
1:                    0.776750   0.0666667           HIGH       7.687533
2:                    1.000000   0.0512821           HIGH      13.064949
3:                    0.797642   0.4871790           HIGH       5.487833
4:                    1.000000   0.4564100           HIGH       7.687347
5:                    0.940012   0.4717950           HIGH       6.174796
6:                    0.940012   0.4717950           HIGH       6.174796
             trait          interval
            &lt;char&gt;            &lt;char&gt;
1:    length_2_4_D 13865698-15025162
2: length_Aldicarb   7790159-8989539
3: length_Aldicarb 16474675-17180844
4: length_Aldicarb 16474675-17180844
5: length_Aldicarb 16474675-17180844
6: length_Aldicarb 16474675-17180844</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>n_row_after <span class="ot">&lt;-</span> <span class="fu">nrow</span>(candidate_table)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(n_row_after)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 179</code></pre>
</div>
</div></div>
</section>
<section id="finemapping-data" class="level2">
<h2 class="anchored" data-anchor-id="finemapping-data">Finemapping data</h2>
<p>Load all of the finemapping data so that we can merge the beta coefficients into the candidate table.</p>
<p>First define the functions to load the finemappings</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [5]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Helper functions to extract information from the finemap file names</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>get_trait_value <span class="ot">&lt;-</span> <span class="cf">function</span>(fn_name) {</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get trait value from file name (keeping the length_ prefix)</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  trait_value <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(fn_name, <span class="st">"length_[^.]+"</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(trait_value)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>get_chromosome <span class="ot">&lt;-</span> <span class="cf">function</span>(fn_name) {</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get chromosome from file name</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  chromosome <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(fn_name, <span class="st">"</span><span class="sc">\\</span><span class="st">b(I|II|III|IV|V|X)</span><span class="sc">\\</span><span class="st">b"</span>)</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(chromosome)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>get_positions <span class="ot">&lt;-</span> <span class="cf">function</span>(fn_name) {</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># split the filename by period</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  parts <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_split</span>(fn_name, <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>)[[<span class="dv">1</span>]]</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>  startPOS <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(parts[<span class="dv">3</span>])</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  stopPOS <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(parts[<span class="dv">4</span>])</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">startPOS =</span> startPOS, <span class="at">stopPOS =</span> stopPOS))</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Main function</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>aggregate_finemappings <span class="ot">&lt;-</span> <span class="cf">function</span>(finemap_dir) {</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>  finemap_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">path =</span> finemap_dir,</span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">pattern =</span> <span class="st">"finemap_inbred.(inbred|loco).fastGWA"</span>,</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">full.names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">recursive =</span> <span class="cn">TRUE</span></span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>    .[<span class="fu">basename</span>(.) <span class="sc">%&gt;%</span> stringr<span class="sc">::</span><span class="fu">str_detect</span>(<span class="st">"^length"</span>)]</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a>  finemap_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">trait_value =</span> <span class="fu">character</span>(),</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">CHROM =</span> <span class="fu">character</span>(),</span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">startPOS =</span> <span class="fu">numeric</span>(),</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">endPOS =</span> <span class="fu">numeric</span>(),</span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">file_name =</span> <span class="fu">character</span>(),</span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">file_path =</span> <span class="fu">character</span>(),</span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb27-40"><a href="#cb27-40" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb27-41"><a href="#cb27-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-42"><a href="#cb27-42" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (fn <span class="cf">in</span> finemap_files) {</span>
<span id="cb27-43"><a href="#cb27-43" aria-hidden="true" tabindex="-1"></a>    fn_name <span class="ot">&lt;-</span> <span class="fu">basename</span>(fn)</span>
<span id="cb27-44"><a href="#cb27-44" aria-hidden="true" tabindex="-1"></a>    trait_value <span class="ot">&lt;-</span> <span class="fu">get_trait_value</span>(fn_name)</span>
<span id="cb27-45"><a href="#cb27-45" aria-hidden="true" tabindex="-1"></a>    chromosome <span class="ot">&lt;-</span> <span class="fu">get_chromosome</span>(fn_name)</span>
<span id="cb27-46"><a href="#cb27-46" aria-hidden="true" tabindex="-1"></a>    positions <span class="ot">&lt;-</span> <span class="fu">get_positions</span>(fn_name)</span>
<span id="cb27-47"><a href="#cb27-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-48"><a href="#cb27-48" aria-hidden="true" tabindex="-1"></a>    finemap_df <span class="ot">&lt;-</span> finemap_df <span class="sc">%&gt;%</span></span>
<span id="cb27-49"><a href="#cb27-49" aria-hidden="true" tabindex="-1"></a>      <span class="fu">add_row</span>(</span>
<span id="cb27-50"><a href="#cb27-50" aria-hidden="true" tabindex="-1"></a>        <span class="at">trait_value =</span> trait_value,</span>
<span id="cb27-51"><a href="#cb27-51" aria-hidden="true" tabindex="-1"></a>        <span class="at">CHROM =</span> chromosome,</span>
<span id="cb27-52"><a href="#cb27-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">startPOS =</span> positions<span class="sc">$</span>startPOS,</span>
<span id="cb27-53"><a href="#cb27-53" aria-hidden="true" tabindex="-1"></a>        <span class="at">endPOS =</span> positions<span class="sc">$</span>stopPOS,</span>
<span id="cb27-54"><a href="#cb27-54" aria-hidden="true" tabindex="-1"></a>        <span class="at">file_name =</span> fn_name,</span>
<span id="cb27-55"><a href="#cb27-55" aria-hidden="true" tabindex="-1"></a>        <span class="at">file_path =</span> fn</span>
<span id="cb27-56"><a href="#cb27-56" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb27-57"><a href="#cb27-57" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb27-58"><a href="#cb27-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-59"><a href="#cb27-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(finemap_df)</span>
<span id="cb27-60"><a href="#cb27-60" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
<div class="cell-container"><div class="cell-decorator"><pre>In [6]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>inbred_finemap_dir <span class="ot">&lt;-</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  glue<span class="sc">::</span><span class="fu">glue</span>(<span class="st">"{ns_folder}/INBRED/Fine_Mappings/Data"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>inbred_finemap_df <span class="ot">&lt;-</span> <span class="fu">aggregate_finemappings</span>(inbred_finemap_dir)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(inbred_finemap_df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      trait_value CHROM startPOS   endPOS
1    length_2_4_D     V 13865698 15025162
2 length_Aldicarb   III  7790159  8989539
3 length_Aldicarb    IV 12966415 13439670
4 length_Aldicarb    IV 16474675 17180844
5 length_Aldicarb    IV  2179845  2507551
6 length_Aldicarb     X 13002956 13505962
                                                           file_name
1     length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2  length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3 length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA
4 length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
5   length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA
6  length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA
                                                                                                                               file_path
1     data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA
4 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
5   data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA
6  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA</code></pre>
</div>
</div></div>
<p>This data frame contains the paths to all of the inbred finemapping results. Next we will filter to just the finemappings that correspond to the candidate genes in our candidate table and then read in those finemapping results into a single data frame.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [7]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a column in the finemap df that matches the interval in the candidate table</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>inbred_finemap_df_m <span class="ot">&lt;-</span> inbred_finemap_df <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">interval =</span> <span class="fu">paste0</span>(startPOS, <span class="st">"-"</span>, endPOS))</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(inbred_finemap_df_m)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      trait_value CHROM startPOS   endPOS
1    length_2_4_D     V 13865698 15025162
2 length_Aldicarb   III  7790159  8989539
3 length_Aldicarb    IV 12966415 13439670
4 length_Aldicarb    IV 16474675 17180844
5 length_Aldicarb    IV  2179845  2507551
6 length_Aldicarb     X 13002956 13505962
                                                           file_name
1     length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2  length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3 length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA
4 length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
5   length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA
6  length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA
                                                                                                                               file_path
1     data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.12966415.13439670.finemap_inbred.inbred.fastGWA
4 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
5   data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.2179845.2507551.finemap_inbred.inbred.fastGWA
6  data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA
           interval
1 13865698-15025162
2   7790159-8989539
3 12966415-13439670
4 16474675-17180844
5   2179845-2507551
6 13002956-13505962</code></pre>
</div>
</div></div>
<p>Create a table of unique trait and interval combinations from the candidate table</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [8]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>candidate_intervals <span class="ot">&lt;-</span> candidate_table <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(trait, interval) <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(candidate_intervals)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     trait          interval
                    &lt;char&gt;            &lt;char&gt;
1:            length_2_4_D 13865698-15025162
2:         length_Aldicarb   7790159-8989539
3:         length_Aldicarb 16474675-17180844
4:         length_Aldicarb 13002956-13505962
5: length_Arsenic_trioxide 13521931-14347284
6: length_Arsenic_trioxide     28947-2527992</code></pre>
</div>
</div></div>
<p>Filter the finemapping data frame to only include files that match the candidate intervals</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [9]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>filtered_finemap_df <span class="ot">&lt;-</span> inbred_finemap_df_m <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">inner_join</span>(candidate_intervals, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"trait_value"</span> <span class="ot">=</span> <span class="st">"trait"</span>, <span class="st">"interval"</span> <span class="ot">=</span> <span class="st">"interval"</span>))</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filtered_finemap_df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>              trait_value CHROM startPOS   endPOS
1            length_2_4_D     V 13865698 15025162
2         length_Aldicarb   III  7790159  8989539
3         length_Aldicarb    IV 16474675 17180844
4         length_Aldicarb     X 13002956 13505962
5 length_Arsenic_trioxide    II 13521931 14347284
6 length_Arsenic_trioxide   III    28947  2527992
                                                                   file_name
1             length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2          length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3         length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
4          length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA
5 length_Arsenic_trioxide.II.13521931.14347284.finemap_inbred.inbred.fastGWA
6    length_Arsenic_trioxide.III.28947.2527992.finemap_inbred.inbred.fastGWA
                                                                                                                                       file_path
1             data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_2_4_D.V.13865698.15025162.finemap_inbred.inbred.fastGWA
2          data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.III.7790159.8989539.finemap_inbred.inbred.fastGWA
3         data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.IV.16474675.17180844.finemap_inbred.inbred.fastGWA
4          data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Aldicarb.X.13002956.13505962.finemap_inbred.inbred.fastGWA
5 data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Arsenic_trioxide.II.13521931.14347284.finemap_inbred.inbred.fastGWA
6    data/processed/20231116_Analysis_NemaScan/INBRED/Fine_Mappings/Data/length_Arsenic_trioxide.III.28947.2527992.finemap_inbred.inbred.fastGWA
           interval
1 13865698-15025162
2   7790159-8989539
3 16474675-17180844
4 13002956-13505962
5 13521931-14347284
6     28947-2527992</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(filtered_finemap_df)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31</code></pre>
</div>
</div></div>
<p>Read in all of the filtered finemapping results and combine them into a single data frame</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [10]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>combine_finemapping_files <span class="ot">&lt;-</span> <span class="cf">function</span>(file_df) {</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  combined_df <span class="ot">&lt;-</span> file_df <span class="sc">%&gt;%</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">data =</span> <span class="fu">list</span>(data.table<span class="sc">::</span><span class="fu">fread</span>(file_path) <span class="sc">%&gt;%</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">trait =</span> trait_value))</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">unnest</span>(<span class="at">cols =</span> <span class="fu">c</span>(data))</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(combined_df)</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>filtered_inbred_finemap_data <span class="ot">&lt;-</span> <span class="fu">combine_finemapping_files</span>(filtered_finemap_df)</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filtered_inbred_finemap_data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
  trait_value  CHROM startPOS   endPOS file_name  file_path interval   CHR SNP  
  &lt;chr&gt;        &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;
1 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
2 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
3 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
4 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
5 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
6 length_2_4_D V     13865698 15025162 length_2_… data/pro… 1386569…     5 5:13…
# ℹ 9 more variables: POS &lt;int&gt;, A1 &lt;chr&gt;, A2 &lt;chr&gt;, N &lt;int&gt;, AF1 &lt;dbl&gt;,
#   BETA &lt;dbl&gt;, SE &lt;dbl&gt;, P &lt;dbl&gt;, trait &lt;chr&gt;</code></pre>
</div>
</div></div>
<p>Now format and pull the relevant columns from the combined finemapping data. Pulling the SNP genotypes (A1, A2) from the finemapping output. Per <a href="https://yanglab.westlake.edu.cn/software/gcta/#fastGWA">GCTA inbred output docs</a></p>
<div class="cell-container"><div class="cell-decorator"><pre>In [11]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>finemap_snvs <span class="ot">&lt;-</span> filtered_inbred_finemap_data <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(trait, SNP, BETA, A1, A2) <span class="sc">%&gt;%</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">MARKER =</span> SNP,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">beta_coefficient =</span> BETA,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">effect_allele =</span> A1,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">other_allele =</span> A2</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">MARKER =</span> <span class="fu">sapply</span>(MARKER, convert_marker, <span class="at">to_marker =</span> <span class="st">"RM"</span>, <span class="at">from_marker =</span> <span class="st">"NUM"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">MARKER =</span> stringr<span class="sc">::</span><span class="fu">str_replace_all</span>(MARKER, <span class="st">":"</span>, <span class="st">"_"</span>))</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(finemap_snvs)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  trait        MARKER     beta_coefficient effect_allele other_allele
  &lt;chr&gt;        &lt;chr&gt;                 &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;       
1 length_2_4_D V_13865698            0.493 G             T           
2 length_2_4_D V_13865823            5.33  A             C           
3 length_2_4_D V_13865980            2.87  A             G           
4 length_2_4_D V_13867914            3.67  C             T           
5 length_2_4_D V_13868694           -3.89  T             C           
6 length_2_4_D V_13869222            8.62  G             A           </code></pre>
</div>
</div></div>
</section>
</section>
<section id="main" class="level1">
<h1>Main</h1>
<p>Join the candidate table with the finemapping SNV data to add the beta coefficients and allele information</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [12]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>candidate_table_with_betas <span class="ot">&lt;-</span> candidate_table <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(finemap_snvs, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"trait"</span>, <span class="st">"MARKER"</span>))</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(candidate_table_with_betas)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>        MARKER  CHROM      POS GENE_NAME       WBGeneID PEAK_MARKER
        &lt;char&gt; &lt;char&gt;    &lt;int&gt;    &lt;char&gt;         &lt;char&gt;      &lt;char&gt;
1:  V_14554027      V 14554027   C53A5.1 WBGene00008270  5:14453636
2: III_8311919    III  8311919   F10E9.3 WBGene00017355   3:8311919
3: IV_16629576     IV 16629576  Y51H4A.9 WBGene00194924  4:16631628
4: IV_16738382     IV 16738382  Y51H4A.2 WBGene00013118  4:16631628
5: IV_16867801     IV 16867801  T06A10.1 WBGene00235300  4:16631628
6: IV_16989319     IV 16989319 Y116A8C.1 WBGene00013791  4:16631628
   VARIANT_LD_WITH_PEAK_MARKER MAF_variant VARIANT_IMPACT VARIANT_LOG10p
                         &lt;num&gt;       &lt;num&gt;         &lt;char&gt;          &lt;num&gt;
1:                    0.776750   0.0666667           HIGH       7.687533
2:                    1.000000   0.0512821           HIGH      13.064949
3:                    0.797642   0.4871790           HIGH       5.487833
4:                    1.000000   0.4564100           HIGH       7.687347
5:                    0.940012   0.4717950           HIGH       6.174796
6:                    0.940012   0.4717950           HIGH       6.174796
             trait          interval beta_coefficient effect_allele
            &lt;char&gt;            &lt;char&gt;            &lt;num&gt;        &lt;char&gt;
1:    length_2_4_D 13865698-15025162          16.5498             A
2: length_Aldicarb   7790159-8989539         -54.6711             T
3: length_Aldicarb 16474675-17180844          16.5350             A
4: length_Aldicarb 16474675-17180844         -20.9067             G
5: length_Aldicarb 16474675-17180844         -18.4316             T
6: length_Aldicarb 16474675-17180844         -18.4316             G
   other_allele
         &lt;char&gt;
1:            T
2:            C
3:            T
4:            T
5:            G
6:            A</code></pre>
</div>
</div></div>
</section>
<section id="outputs" class="level1">
<h1>Outputs</h1>
<p>Save the candidate table with beta coefficients</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [13]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>candidate_table_betas_outfn <span class="ot">&lt;-</span> <span class="st">"data/processed/interval_genes/candidate_genes/inbred_candidate_genes_with_betas.tsv.gz"</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save_compressed_tsv</span>(candidate_table_with_betas, candidate_table_betas_outfn)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
</section>
     </main>
<!-- /main column -->  <script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>  </div> <!-- /content --> 
  
</body></html>