<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
    <meta charset="utf-8">
    <meta name="generator" content="quarto-1.8.23">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


    <title>Analysis 11: Calculate Region HDR Overlaps</title>
    <style>
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      div.columns{display: flex; gap: min(4vw, 1.5em);}
      div.column{flex: auto; overflow-x: auto;}
      div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
      ul.task-list{list-style: none;}
      ul.task-list li input[type="checkbox"] {
        width: 0.8em;
        margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
        vertical-align: middle;
      }
      /* CSS for syntax highlighting */
      html { -webkit-text-size-adjust: 100%; }
      pre > code.sourceCode { white-space: pre; position: relative; }
      pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
      pre > code.sourceCode > span:empty { height: 1.2em; }
      .sourceCode { overflow: visible; }
      code.sourceCode > span { color: inherit; text-decoration: inherit; }
      div.sourceCode { margin: 1em 0; }
      pre.sourceCode { margin: 0; }
      @media screen {
      div.sourceCode { overflow: auto; }
      }
      @media print {
      pre > code.sourceCode { white-space: pre-wrap; }
      pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
      }
      pre.numberSource code
        { counter-reset: source-line 0; }
      pre.numberSource code > span
        { position: relative; left: -4em; counter-increment: source-line; }
      pre.numberSource code > span > a:first-child::before
        { content: counter(source-line);
          position: relative; left: -1em; text-align: right; vertical-align: baseline;
          border: none; display: inline-block;
          -webkit-touch-callout: none; -webkit-user-select: none;
          -khtml-user-select: none; -moz-user-select: none;
          -ms-user-select: none; user-select: none;
          padding: 0 4px; width: 4em;
        }
      pre.numberSource { margin-left: 3em;  padding-left: 4px; }
      div.sourceCode
        {   }
      @media screen {
      pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
      }
    </style>

    <style>
      body.hypothesis-enabled #quarto-embed-header {
        padding-right: 36px;
      }

      #quarto-embed-header {
        height: 3em;
        width: 100%;
        display: flex;
        justify-content: space-between;
        align-items: center;
        border-bottom: solid 1px;
      }

      #quarto-embed-header h6 {
        font-size: 1.1em;
        padding-top: 0.6em;
        margin-left: 1em;
        margin-right: 1em;
        font-weight: 400;
      }

      #quarto-embed-header a.quarto-back-link,
      #quarto-embed-header a.quarto-download-embed {
        font-size: 0.8em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 1em;
        margin-right: 1em;
      }

      .quarto-back-container {
        padding-left: 0.5em;
        display: flex;
      }

      .headroom {
          will-change: transform;
          transition: transform 200ms linear;
      }

      .headroom--pinned {
          transform: translateY(0%);
      }

      .headroom--unpinned {
          transform: translateY(-100%);
      }      
    </style>

    <script>
    window.document.addEventListener("DOMContentLoaded", function () {

      var header = window.document.querySelector("#quarto-embed-header");
      const titleBannerEl = window.document.querySelector("body > #title-block-header");
      if (titleBannerEl) {
        titleBannerEl.style.paddingTop = header.clientHeight + "px";
      }
      const contentEl = window.document.getElementById('quarto-content');
      for (const child of contentEl.children) {
        child.style.paddingTop = header.clientHeight + "px";
        child.style.marginTop = "1em";
      }

      // Use the article root if the `back` call doesn't work. This isn't perfect
      // but should typically work
      window.quartoBackToArticle = () => {
        var currentUrl = window.location.href;
        window.history.back();
        setTimeout(() => {
            // if location was not changed in 100 ms, then there is no history back
            if(currentUrl === window.location.href){              
                // redirect to site root
                window.location.href = "../index.html";
            }
        }, 100);
      }

      const headroom = new window.Headroom(header, {
        tolerance: 5,
        onPin: function () {
        },
        onUnpin: function () {
        },
      });
      headroom.init();
    });
    </script>

    
<script src="../site_libs/manuscript-notebook/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-1fe81d0376b2c50856e68e651e390326.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a74871fe4945b66d259aafc266475145.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
     
      </head>

  <body class="quarto-notebook quarto-light">
    <div id="quarto-embed-header" class="headroom fixed-top bg-primary">
      
      <a onclick="window.quartoBackToArticle(); return false;" class="btn btn-primary quarto-back-link" href=""><i class="bi bi-caret-left"></i> Back to Article</a>
      <h6><i class="bi bi-journal-code"></i> Analysis 11: Calculate Region HDR Overlaps</h6>

            <a href="../analysis/hdr_calculate_region_overlaps.qmd" class="btn btn-primary quarto-download-embed" download="hdr_calculate_region_overlaps.qmd">Download Source</a>
          </div>

     <header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Analysis 11: Calculate Region HDR Overlaps</h1>
          </div>

    
    <div class="quarto-title-meta-container">
      <div class="quarto-title-meta-column-start">
        
        <div class="quarto-title-meta">

                
          
                
              </div>
      </div>
      <div class="quarto-title-meta-column-end quarto-other-formats-target">
      </div>
    </div>



    <div class="quarto-other-links-text-target">
    </div>  </div>
</header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#inputs" id="toc-inputs" class="nav-link active" data-scroll-target="#inputs">Inputs</a></li>
  <li><a href="#outputs" id="toc-outputs" class="nav-link" data-scroll-target="#outputs">Outputs</a></li>
  <li><a href="#functions" id="toc-functions" class="nav-link" data-scroll-target="#functions">Functions</a></li>
  <li><a href="#main" id="toc-main" class="nav-link" data-scroll-target="#main">Main</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">      

       <p>Based on <code>DivergentRegion_Overlaps.R</code> script from Analysis GitHub repo.</p>
<p>Take the HVR regions present in the mapping panel and measure the overlap of intervals detected for each toxicant.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [1]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4
Loading required package: BiocGenerics

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:lubridate':

    intersect, setdiff, union

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min

Loading required package: S4Vectors

Attaching package: 'S4Vectors'

The following objects are masked from 'package:lubridate':

    second, second&lt;-

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following objects are masked from 'package:base':

    expand.grid, I, unname

Loading required package: IRanges

Attaching package: 'IRanges'

The following object is masked from 'package:lubridate':

    %within%

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce

Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'data.table'

The following object is masked from 'package:GenomicRanges':

    shift

The following object is masked from 'package:IRanges':

    shift

The following objects are masked from 'package:S4Vectors':

    first, second

The following objects are masked from 'package:lubridate':

    hour, isoweek, mday, minute, month, quarter, second, wday, week,
    yday, year

The following objects are masked from 'package:dplyr':

    between, first, last

The following object is masked from 'package:purrr':

    transpose</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>------------------------------------------------------------------------------
You have loaded plyr after dplyr - this is likely to cause problems.
If you need functions from both plyr and dplyr, please load plyr first, then dplyr:
library(plyr); library(dplyr)
------------------------------------------------------------------------------

Attaching package: 'plyr'

The following object is masked from 'package:IRanges':

    desc

The following object is masked from 'package:S4Vectors':

    rename

The following objects are masked from 'package:dplyr':

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize

The following object is masked from 'package:purrr':

    compact</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>today <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">Sys.time</span>(), <span class="st">"%Y%m%d_%H%M"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
<section id="inputs" class="level1">
<h1>Inputs</h1>
<ul>
<li><code>hd_regions_file</code>: path to the <code>hvr_gwas_strains.tsv</code> file generated by <code>analysis/hdr_summarize_strain.qmd</code></li>
<li><code>ns_folder</code>: path to the nemascan analysis folder</li>
</ul>
<div class="cell-container"><div class="cell-decorator"><pre>In [2]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># path to HDR regions file for GWAS strains from analysis/hdr_summarize_strain.qmd</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>hd_regions_file <span class="ot">&lt;-</span> <span class="st">"data/processed/HDR_overlap/hvr_gwas_strains.tsv"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Path to folder with NemaScan results</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>ns_folder <span class="ot">&lt;-</span> <span class="st">"data/processed/20231116_Analysis_NemaScan"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
</section>
<section id="outputs" class="level1">
<h1>Outputs</h1>
<p>In the output directory (<code>data/processed/HVR_Overlap/All_QTL_HD_overlap</code>) there is a file generated for each trait (e.g <code>length_2_4_D_HD_region_overlap.tsv</code>). Each row in the file represented a qtl detected for that trait. These files are used later in the <code>triage</code> scripts.</p>
<div class="cell-container"><div class="cell-decorator"><pre>In [3]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> glue<span class="sc">::</span><span class="fu">glue</span>(<span class="st">"data/processed/HVR_Overlap/All_QTL_HD_overlap"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#check if the outdir exists, if not create it</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(outdir)) {</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(outdir)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(outdir): cannot create dir
'data/processed/HVR_Overlap/All_QTL_HD_overlap', reason 'No such file or
directory'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create specific output directories for inbred mappings</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>inbred_out_dir <span class="ot">&lt;-</span> glue<span class="sc">::</span><span class="fu">glue</span>(<span class="st">"{outdir}/inbred"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(inbred_out_dir)) {</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(inbred_out_dir)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(inbred_out_dir): cannot create dir
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred', reason 'No such file or
directory'</code></pre>
</div>
</div></div>
</section>
<section id="functions" class="level1">
<h1>Functions</h1>
<div class="cell-container"><div class="cell-decorator"><pre>In [4]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>create_proc_mapping <span class="ot">&lt;-</span> <span class="cf">function</span>(mapping_file) {</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    mapping <span class="ot">&lt;-</span> <span class="fu">fread</span>(mapping_file)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    mapping <span class="ot">&lt;-</span> mapping <span class="sc">%&gt;%</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(peak_id))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    n_peak_markers <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(mapping<span class="sc">$</span>peak_id))</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Number of markers with a peak ID"</span>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(n_peak_markers)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (n_peak_markers <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>        mapping_list <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    mapping_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n_peak_markers) {</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>        fil_mapping <span class="ot">&lt;-</span> mapping <span class="sc">%&gt;%</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(peak_id <span class="sc">==</span> i)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>        peak_marker <span class="ot">&lt;-</span> <span class="fu">unique</span>(fil_mapping<span class="sc">$</span>marker)</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>        ref_strains <span class="ot">&lt;-</span> fil_mapping <span class="sc">%&gt;%</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(allele <span class="sc">==</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>            <span class="fu">distinct</span>(strain) <span class="sc">%&gt;%</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>            <span class="fu">pull</span>(strain)</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>        alt_strains <span class="ot">&lt;-</span> fil_mapping <span class="sc">%&gt;%</span></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(allele <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>            <span class="fu">distinct</span>(strain) <span class="sc">%&gt;%</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>            <span class="fu">pull</span>(strain)</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>        single_peak_df <span class="ot">&lt;-</span> fil_mapping  <span class="sc">%&gt;%</span> </span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>            <span class="fu">select</span>(CHROM, startPOS, endPOS) <span class="sc">%&gt;%</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>            <span class="fu">distinct</span>()</span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Create a genomic ranges object for the peak</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>        peak_gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> single_peak_df<span class="sc">$</span>CHROM,</span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>                           <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> single_peak_df<span class="sc">$</span>startPOS,</span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">end =</span> single_peak_df<span class="sc">$</span>endPOS),</span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>        peak_data <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">ref_strains =</span> ref_strains, <span class="at">alt_strains =</span> alt_strains, <span class="at">peak_gr =</span> peak_gr)</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>        mapping_list[[peak_marker]] <span class="ot">&lt;-</span> peak_data</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(mapping_list)</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to create a genomic ranges object for the HD regions for the strain set</span></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>create_hd_gr <span class="ot">&lt;-</span> <span class="cf">function</span>(hd_file) {</span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a>    hd_df <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(</span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>        hd_file,</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>        <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"stop"</span>, <span class="st">"hd_region"</span>, <span class="st">"strains"</span>, <span class="st">"n_strains"</span>)</span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>    hd_gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> hd_df<span class="sc">$</span>chrom,</span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>                     <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> hd_df<span class="sc">$</span>start,</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">end =</span> hd_df<span class="sc">$</span>end),</span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>                     <span class="at">hd_region =</span> hd_df<span class="sc">$</span>hd_region,</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>                     <span class="at">strains =</span> hd_df<span class="sc">$</span>strains,</span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n_strains =</span> hd_df<span class="sc">$</span>n_strains</span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>                     )</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(hd_gr)</span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a><span class="co">#Create a function to check if a peak overlaps with a HD region</span></span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a>strains_with_hd_overlap <span class="ot">&lt;-</span> <span class="cf">function</span>(peak_gr, hd_gr) {</span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a>    <span class="co">#return only the ranges in the hd_gr that overlap with the peak_gr</span></span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a>    hd_overlap <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(hd_gr, peak_gr)</span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a>    n_hd_overlaps <span class="ot">&lt;-</span> <span class="fu">length</span>(hd_overlap)</span>
<span id="cb15-75"><a href="#cb15-75" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>( <span class="st">"Number of HD regions that overlap with a QTL region"</span>)</span>
<span id="cb15-76"><a href="#cb15-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(n_hd_overlaps)</span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a>    <span class="co">#return the hd_overlap object as a data frame</span></span>
<span id="cb15-79"><a href="#cb15-79" aria-hidden="true" tabindex="-1"></a>    hd_overlap_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(hd_overlap) <span class="sc">%&gt;%</span> </span>
<span id="cb15-80"><a href="#cb15-80" aria-hidden="true" tabindex="-1"></a>    <span class="co">#split the strains column into a list</span></span>
<span id="cb15-81"><a href="#cb15-81" aria-hidden="true" tabindex="-1"></a>        <span class="fu">separate_rows</span>(strains, <span class="at">sep =</span> <span class="st">","</span>) </span>
<span id="cb15-82"><a href="#cb15-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-83"><a href="#cb15-83" aria-hidden="true" tabindex="-1"></a>    hd_overlap_strains <span class="ot">&lt;-</span> hd_overlap_df <span class="sc">%&gt;%</span></span>
<span id="cb15-84"><a href="#cb15-84" aria-hidden="true" tabindex="-1"></a>        <span class="fu">select</span>(strains) <span class="sc">%&gt;%</span></span>
<span id="cb15-85"><a href="#cb15-85" aria-hidden="true" tabindex="-1"></a>        <span class="fu">distinct</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb15-86"><a href="#cb15-86" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pull</span>(strains)</span>
<span id="cb15-87"><a href="#cb15-87" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-88"><a href="#cb15-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(hd_overlap_strains)</span>
<span id="cb15-89"><a href="#cb15-89" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-90"><a href="#cb15-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-91"><a href="#cb15-91" aria-hidden="true" tabindex="-1"></a><span class="co">#Load in the mapping data</span></span>
<span id="cb15-92"><a href="#cb15-92" aria-hidden="true" tabindex="-1"></a><span class="co">#example_mapping_file &lt;- "/projects/b1059/projects/Tim/2021_GWA_manuscript/data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carboxin_AGGREGATE_mapping_inbred.tsv" </span></span>
<span id="cb15-93"><a href="#cb15-93" aria-hidden="true" tabindex="-1"></a><span class="co">#mapping_list &lt;- create_proc_mapping(example_mapping_file)</span></span>
<span id="cb15-94"><a href="#cb15-94" aria-hidden="true" tabindex="-1"></a><span class="co">#head(mapping_list)</span></span>
<span id="cb15-95"><a href="#cb15-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-96"><a href="#cb15-96" aria-hidden="true" tabindex="-1"></a><span class="co">#Load in the HD regions data</span></span>
<span id="cb15-97"><a href="#cb15-97" aria-hidden="true" tabindex="-1"></a><span class="co">#hd_regions_file &lt;- "data/20231121_divergent_regions_strain.bed"</span></span>
<span id="cb15-98"><a href="#cb15-98" aria-hidden="true" tabindex="-1"></a><span class="co">#hd_gr &lt;- create_hd_gr(hd_regions_file)</span></span>
<span id="cb15-99"><a href="#cb15-99" aria-hidden="true" tabindex="-1"></a><span class="co">#head(hd_gr)</span></span>
<span id="cb15-100"><a href="#cb15-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-101"><a href="#cb15-101" aria-hidden="true" tabindex="-1"></a>create_trait_overlaps <span class="ot">&lt;-</span> <span class="cf">function</span>(mapping_file, hd_regions_file, outdir){</span>
<span id="cb15-102"><a href="#cb15-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-103"><a href="#cb15-103" aria-hidden="true" tabindex="-1"></a><span class="co">#Create mapping files obj</span></span>
<span id="cb15-104"><a href="#cb15-104" aria-hidden="true" tabindex="-1"></a>mapping_list <span class="ot">&lt;-</span> <span class="fu">create_proc_mapping</span>(mapping_file)</span>
<span id="cb15-105"><a href="#cb15-105" aria-hidden="true" tabindex="-1"></a><span class="co">#if mapping list is null, return null</span></span>
<span id="cb15-106"><a href="#cb15-106" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(mapping_list)) {</span>
<span id="cb15-107"><a href="#cb15-107" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"No markers with peak IDs"</span>)</span>
<span id="cb15-108"><a href="#cb15-108" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-109"><a href="#cb15-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-110"><a href="#cb15-110" aria-hidden="true" tabindex="-1"></a><span class="co">#Create HD regions obj</span></span>
<span id="cb15-111"><a href="#cb15-111" aria-hidden="true" tabindex="-1"></a>hd_gr <span class="ot">&lt;-</span> <span class="fu">create_hd_gr</span>(hd_regions_file)</span>
<span id="cb15-112"><a href="#cb15-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-113"><a href="#cb15-113" aria-hidden="true" tabindex="-1"></a><span class="co">#Iterate through the mapping list and check for HD region overlaps</span></span>
<span id="cb15-114"><a href="#cb15-114" aria-hidden="true" tabindex="-1"></a>hd_overlap_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-115"><a href="#cb15-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-116"><a href="#cb15-116" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(mapping_list)) {</span>
<span id="cb15-117"><a href="#cb15-117" aria-hidden="true" tabindex="-1"></a>    peak_data <span class="ot">&lt;-</span> mapping_list[[i]]</span>
<span id="cb15-118"><a href="#cb15-118" aria-hidden="true" tabindex="-1"></a>    <span class="co">#get the marker id for the peak from the name of the list element</span></span>
<span id="cb15-119"><a href="#cb15-119" aria-hidden="true" tabindex="-1"></a>    peak_marker <span class="ot">&lt;-</span> <span class="fu">names</span>(mapping_list[i])</span>
<span id="cb15-120"><a href="#cb15-120" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(peak_marker)</span>
<span id="cb15-121"><a href="#cb15-121" aria-hidden="true" tabindex="-1"></a>    peak_gr <span class="ot">&lt;-</span> peak_data<span class="sc">$</span>peak_gr</span>
<span id="cb15-122"><a href="#cb15-122" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-123"><a href="#cb15-123" aria-hidden="true" tabindex="-1"></a>    <span class="co">#calculate the number of strains with either ref or alt status at the peak marker</span></span>
<span id="cb15-124"><a href="#cb15-124" aria-hidden="true" tabindex="-1"></a>    n_strains_ref <span class="ot">&lt;-</span> <span class="fu">length</span>(peak_data<span class="sc">$</span>ref_strains)</span>
<span id="cb15-125"><a href="#cb15-125" aria-hidden="true" tabindex="-1"></a>    n_strains_alt <span class="ot">&lt;-</span> <span class="fu">length</span>(peak_data<span class="sc">$</span>alt_strains)</span>
<span id="cb15-126"><a href="#cb15-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-127"><a href="#cb15-127" aria-hidden="true" tabindex="-1"></a>    total_strains <span class="ot">&lt;-</span> n_strains_ref <span class="sc">+</span> n_strains_alt</span>
<span id="cb15-128"><a href="#cb15-128" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-129"><a href="#cb15-129" aria-hidden="true" tabindex="-1"></a>    hd_overlap_strains <span class="ot">&lt;-</span> <span class="fu">strains_with_hd_overlap</span>(peak_gr, hd_gr)</span>
<span id="cb15-130"><a href="#cb15-130" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-131"><a href="#cb15-131" aria-hidden="true" tabindex="-1"></a>    n_strains_ref_overlap <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(peak_data<span class="sc">$</span>ref_strains, hd_overlap_strains))</span>
<span id="cb15-132"><a href="#cb15-132" aria-hidden="true" tabindex="-1"></a>    n_strains_alt_overlap <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(peak_data<span class="sc">$</span>alt_strains, hd_overlap_strains))</span>
<span id="cb15-133"><a href="#cb15-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-134"><a href="#cb15-134" aria-hidden="true" tabindex="-1"></a>    hd_overlap_list[[peak_marker]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb15-135"><a href="#cb15-135" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">peak_marker =</span> peak_marker,</span>
<span id="cb15-136"><a href="#cb15-136" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">n_strains_ref =</span> n_strains_ref,</span>
<span id="cb15-137"><a href="#cb15-137" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">n_strains_alt =</span> n_strains_alt,</span>
<span id="cb15-138"><a href="#cb15-138" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">total_strains =</span> total_strains,</span>
<span id="cb15-139"><a href="#cb15-139" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">n_strains_ref_overlap =</span> n_strains_ref_overlap,</span>
<span id="cb15-140"><a href="#cb15-140" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">n_strains_alt_overlap =</span> n_strains_alt_overlap,</span>
<span id="cb15-141"><a href="#cb15-141" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">prop_ref_overlap =</span> n_strains_ref_overlap<span class="sc">/</span>n_strains_ref,</span>
<span id="cb15-142"><a href="#cb15-142" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">prop_alt_overlap =</span> n_strains_alt_overlap<span class="sc">/</span>n_strains_alt</span>
<span id="cb15-143"><a href="#cb15-143" aria-hidden="true" tabindex="-1"></a>                                           )</span>
<span id="cb15-144"><a href="#cb15-144" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-145"><a href="#cb15-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-146"><a href="#cb15-146" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hd_overlap_list)</span>
<span id="cb15-147"><a href="#cb15-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-148"><a href="#cb15-148" aria-hidden="true" tabindex="-1"></a>hd_overlap_df <span class="ot">&lt;-</span> <span class="fu">ldply</span>(hd_overlap_list, data.frame)</span>
<span id="cb15-149"><a href="#cb15-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-150"><a href="#cb15-150" aria-hidden="true" tabindex="-1"></a><span class="co">#extract the trait name from the mapping file name</span></span>
<span id="cb15-151"><a href="#cb15-151" aria-hidden="true" tabindex="-1"></a>get_traitname <span class="ot">&lt;-</span> <span class="cf">function</span>(fn){</span>
<span id="cb15-152"><a href="#cb15-152" aria-hidden="true" tabindex="-1"></a>    bn <span class="ot">&lt;-</span> <span class="fu">basename</span>(fn)</span>
<span id="cb15-153"><a href="#cb15-153" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_detect</span>(bn, <span class="st">"CV_length"</span>)){</span>
<span id="cb15-154"><a href="#cb15-154" aria-hidden="true" tabindex="-1"></a>        <span class="co">#extract everthing before _AGGREGATE</span></span>
<span id="cb15-155"><a href="#cb15-155" aria-hidden="true" tabindex="-1"></a>        traitname <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(bn, <span class="st">".*(?=_AGGREGATE)"</span>)</span>
<span id="cb15-156"><a href="#cb15-156" aria-hidden="true" tabindex="-1"></a>        traitname <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(traitname, <span class="st">"processed_"</span>)</span>
<span id="cb15-157"><a href="#cb15-157" aria-hidden="true" tabindex="-1"></a>    }<span class="cf">else</span>{</span>
<span id="cb15-158"><a href="#cb15-158" aria-hidden="true" tabindex="-1"></a>        <span class="co">#extract everything before _AGGREGATE </span></span>
<span id="cb15-159"><a href="#cb15-159" aria-hidden="true" tabindex="-1"></a>        traitname <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(bn, <span class="st">".*(?=_AGGREGATE)"</span>)</span>
<span id="cb15-160"><a href="#cb15-160" aria-hidden="true" tabindex="-1"></a>        traitname <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(traitname, <span class="st">"processed_"</span>)</span>
<span id="cb15-161"><a href="#cb15-161" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-162"><a href="#cb15-162" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(traitname)</span>
<span id="cb15-163"><a href="#cb15-163" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-164"><a href="#cb15-164" aria-hidden="true" tabindex="-1"></a>trait_name <span class="ot">&lt;-</span> <span class="fu">get_traitname</span>(mapping_file)</span>
<span id="cb15-165"><a href="#cb15-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-166"><a href="#cb15-166" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(trait_name)</span>
<span id="cb15-167"><a href="#cb15-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-168"><a href="#cb15-168" aria-hidden="true" tabindex="-1"></a>hd_overlap_df<span class="sc">$</span>trait_name <span class="ot">&lt;-</span> trait_name</span>
<span id="cb15-169"><a href="#cb15-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-170"><a href="#cb15-170" aria-hidden="true" tabindex="-1"></a><span class="co">#Add the algoirthm to the data frame</span></span>
<span id="cb15-171"><a href="#cb15-171" aria-hidden="true" tabindex="-1"></a>get_algorithm <span class="ot">&lt;-</span> <span class="cf">function</span>(fn){</span>
<span id="cb15-172"><a href="#cb15-172" aria-hidden="true" tabindex="-1"></a>    bn <span class="ot">&lt;-</span> <span class="fu">basename</span>(fn)</span>
<span id="cb15-173"><a href="#cb15-173" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_detect</span>(bn, <span class="st">"inbred"</span>)){</span>
<span id="cb15-174"><a href="#cb15-174" aria-hidden="true" tabindex="-1"></a>        algorithm <span class="ot">&lt;-</span> <span class="st">"inbred"</span></span>
<span id="cb15-175"><a href="#cb15-175" aria-hidden="true" tabindex="-1"></a>    }<span class="cf">else</span>{</span>
<span id="cb15-176"><a href="#cb15-176" aria-hidden="true" tabindex="-1"></a>        algorithm <span class="ot">&lt;-</span> <span class="st">"loco"</span></span>
<span id="cb15-177"><a href="#cb15-177" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-178"><a href="#cb15-178" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(algorithm)</span>
<span id="cb15-179"><a href="#cb15-179" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-180"><a href="#cb15-180" aria-hidden="true" tabindex="-1"></a>algorithm <span class="ot">&lt;-</span> <span class="fu">get_algorithm</span>(mapping_file)</span>
<span id="cb15-181"><a href="#cb15-181" aria-hidden="true" tabindex="-1"></a>hd_overlap_df<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> algorithm</span>
<span id="cb15-182"><a href="#cb15-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-183"><a href="#cb15-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-184"><a href="#cb15-184" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hd_overlap_df)</span>
<span id="cb15-185"><a href="#cb15-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-186"><a href="#cb15-186" aria-hidden="true" tabindex="-1"></a><span class="co">#write out the data frame with the HD region overlap data using the trait name</span></span>
<span id="cb15-187"><a href="#cb15-187" aria-hidden="true" tabindex="-1"></a><span class="co">#hd_overlap_outfile &lt;- paste0("data/", trait_name, "_HD_region_overlap.tsv")</span></span>
<span id="cb15-188"><a href="#cb15-188" aria-hidden="true" tabindex="-1"></a>hd_overlap_outfile <span class="ot">&lt;-</span> glue<span class="sc">::</span><span class="fu">glue</span>(<span class="st">"{outdir}/{trait_name}_HD_region_overlap.tsv"</span>)</span>
<span id="cb15-189"><a href="#cb15-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-190"><a href="#cb15-190" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if the file exists</span></span>
<span id="cb15-191"><a href="#cb15-191" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(hd_overlap_outfile)) {</span>
<span id="cb15-192"><a href="#cb15-192" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If the file exists, read the existing data</span></span>
<span id="cb15-193"><a href="#cb15-193" aria-hidden="true" tabindex="-1"></a>    existing_data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(hd_overlap_outfile, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb15-194"><a href="#cb15-194" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-195"><a href="#cb15-195" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Append the new data to the existing data</span></span>
<span id="cb15-196"><a href="#cb15-196" aria-hidden="true" tabindex="-1"></a>    combined_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(existing_data, hd_overlap_df)</span>
<span id="cb15-197"><a href="#cb15-197" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb15-198"><a href="#cb15-198" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If the file does not exist, the new data is the combined data</span></span>
<span id="cb15-199"><a href="#cb15-199" aria-hidden="true" tabindex="-1"></a>    combined_data <span class="ot">&lt;-</span> hd_overlap_df</span>
<span id="cb15-200"><a href="#cb15-200" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-201"><a href="#cb15-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-202"><a href="#cb15-202" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the combined data to the file</span></span>
<span id="cb15-203"><a href="#cb15-203" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(combined_data, hd_overlap_outfile, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb15-204"><a href="#cb15-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-205"><a href="#cb15-205" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></div>
</section>
<section id="main" class="level1">
<h1>Main</h1>
<p>The main function executed by this script is <code>create_trait_overlaps()</code> which executes a series of other functions defined in the script.</p>
<ol type="1">
<li>Create a mapping files object (a list with several named elements) with <code>create_proc_mapping()</code>
<ul>
<li>Function loads the mapping file containing the p-values for all markers and filters to just the peak marker for each interval that was defined by NS.</li>
<li>Using the filtered mapping data the function counts a list of strains with the REF or ALT allele peak marker allele. Generates the <code>ref_strains</code> and <code>alt_strains</code> elements of the mapping files object.</li>
<li>Creates a genomic ranges object using data about the interval that peak marker represents. This is the <code>peak_gr</code> element of the mapping files object.</li>
</ul></li>
<li>Create a genomic range object from the HVR bedfile using the <code>create_hd_gr()</code> function.</li>
<li>Identify strains with that have an HVR in the region of the peak marker using the <code>strains_with_hd_overlap()</code> function.</li>
<li>Calculate the proportion of strains with each peak marker allele REF or ALT that have an HVR overlap.</li>
</ol>
<div class="cell-container"><div class="cell-decorator"><pre>In [5]:</pre></div><div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Get the mapping directories</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>inbred_mapping_dir <span class="ot">&lt;-</span> glue<span class="sc">::</span><span class="fu">glue</span>(<span class="st">"{ns_folder}/INBRED/Mapping/Processed"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Get the mapping files</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>inbred_mapping_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">path =</span> inbred_mapping_dir,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pattern =</span> <span class="st">"AGGREGATE_mapping_inbred"</span>,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">recursive =</span> T, <span class="at">full.names =</span> T)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Get rid of CV length files</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>inbred_mapping_files <span class="ot">&lt;-</span> inbred_mapping_files[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"_CV_"</span>, inbred_mapping_files)]</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="co">#Iterate through the mapping files and create the HD region overlap data</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(inbred_mapping_files)) {</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    mapping_file <span class="ot">&lt;-</span> inbred_mapping_files[i]</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(mapping_file)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tryCatch</span>(<span class="fu">create_trait_overlaps</span>(mapping_file, hd_regions_file, inbred_out_dir),</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>                <span class="at">error =</span> <span class="cf">function</span>(e) <span class="fu">print</span>(e)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>    <span class="co">#create_trait_overlaps(mapping_file, hd_regions_file, outdir)</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_2_4_D_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "V:14453636"
[1] "Number of HD regions that overlap with a QTL region"
[1] 10
[1] "length_2_4_D"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_2_4_D_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Aldicarb_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 5
[1] "III:8311919"
[1] "Number of HD regions that overlap with a QTL region"
[1] 0
[1] "IV:2368192"
[1] "Number of HD regions that overlap with a QTL region"
[1] 6
[1] "IV:13208614"
[1] "Number of HD regions that overlap with a QTL region"
[1] 2
[1] "IV:16631628"
[1] "Number of HD regions that overlap with a QTL region"
[1] 29
[1] "X:13252751"
[1] "Number of HD regions that overlap with a QTL region"
[1] 8
[1] "length_Aldicarb"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Aldicarb_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Arsenic_trioxide_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 2
[1] "II:13692527"
[1] "Number of HD regions that overlap with a QTL region"
[1] 76
[1] "III:1642838"
[1] "Number of HD regions that overlap with a QTL region"
[1] 198
[1] "length_Arsenic_trioxide"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Arsenic_trioxide_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Atrazine_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "II:12658388"
[1] "Number of HD regions that overlap with a QTL region"
[1] 36
[1] "length_Atrazine"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Atrazine_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Cadmium_dichloride_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "X:5955567"
[1] "Number of HD regions that overlap with a QTL region"
[1] 14
[1] "length_Cadmium_dichloride"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Cadmium_dichloride_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carbaryl_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "I:2596756"
[1] "Number of HD regions that overlap with a QTL region"
[1] 13
[1] "length_Carbaryl"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Carbaryl_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carboxin_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "III:2527992"
[1] "Number of HD regions that overlap with a QTL region"
[1] 32
[1] "length_Carboxin"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Carboxin_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorfenapyr_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorothalonil_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 6
[1] "I:1741155"
[1] "Number of HD regions that overlap with a QTL region"
[1] 13
[1] "II:5392976"
[1] "Number of HD regions that overlap with a QTL region"
[1] 88
[1] "II:13921653"
[1] "Number of HD regions that overlap with a QTL region"
[1] 63
[1] "III:1455164"
[1] "Number of HD regions that overlap with a QTL region"
[1] 51
[1] "III:2605810"
[1] "Number of HD regions that overlap with a QTL region"
[1] 36
[1] "III:4478684"
[1] "Number of HD regions that overlap with a QTL region"
[1] 6
[1] "length_Chlorothalonil"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Chlorothalonil_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorpyrifos_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Copper_chloride_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_DMSO_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Lead_nitrate_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Malathion_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Mancozeb_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "II:10305644"
[1] "Number of HD regions that overlap with a QTL region"
[1] 7
[1] "length_Mancozeb"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Mancozeb_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Methomyl_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Methyl_mercury_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 8
[1] "I:2994073"
[1] "Number of HD regions that overlap with a QTL region"
[1] 5
[1] "II:626831"
[1] "Number of HD regions that overlap with a QTL region"
[1] 44
[1] "II:2730209"
[1] "Number of HD regions that overlap with a QTL region"
[1] 21
[1] "II:3977000"
[1] "Number of HD regions that overlap with a QTL region"
[1] 32
[1] "III:2461452"
[1] "Number of HD regions that overlap with a QTL region"
[1] 35
[1] "III:3716201"
[1] "Number of HD regions that overlap with a QTL region"
[1] 1
[1] "III:10133612"
[1] "Number of HD regions that overlap with a QTL region"
[1] 7
[1] "V:15476283"
[1] "Number of HD regions that overlap with a QTL region"
[1] 61
[1] "length_Methyl_mercury"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Methyl_mercury_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Nickel_dichloride_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 2
[1] "II:12203935"
[1] "Number of HD regions that overlap with a QTL region"
[1] 8
[1] "V:1486141"
[1] "Number of HD regions that overlap with a QTL region"
[1] 15
[1] "length_Nickel_dichloride"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Nickel_dichloride_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Paraquat_250_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Paraquat_62_5_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 4
[1] "I:3282224"
[1] "Number of HD regions that overlap with a QTL region"
[1] 0
[1] "I:11258815"
[1] "Number of HD regions that overlap with a QTL region"
[1] 11
[1] "II:13675568"
[1] "Number of HD regions that overlap with a QTL region"
[1] 57
[1] "X:2848181"
[1] "Number of HD regions that overlap with a QTL region"
[1] 9
[1] "length_Paraquat_62_5"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Paraquat_62_5_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Propoxur_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "V:5315004"
[1] "Number of HD regions that overlap with a QTL region"
[1] 14
[1] "length_Propoxur"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Propoxur_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Pyraclostrobin_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Silver_nitrate_250_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Silver_nitrate_7_8_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 3
[1] "I:1622740"
[1] "Number of HD regions that overlap with a QTL region"
[1] 17
[1] "II:3016298"
[1] "Number of HD regions that overlap with a QTL region"
[1] 128
[1] "II:11883855"
[1] "Number of HD regions that overlap with a QTL region"
[1] 6
[1] "length_Silver_nitrate_7_8"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Silver_nitrate_7_8_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Triphenyl_phosphate_50_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Triphenyl_phosphate_6_25_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 2
[1] "IV:16034033"
[1] "Number of HD regions that overlap with a QTL region"
[1] 5
[1] "V:7623783"
[1] "Number of HD regions that overlap with a QTL region"
[1] 229
[1] "length_Triphenyl_phosphate_6_25"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Triphenyl_phosphate_6_25_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Water_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 0
&lt;simpleError in mapping_list[[peak_marker]] &lt;- peak_data: attempt to select less than one element in OneIndex&gt;
[1] "data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Zinc_dichloride_AGGREGATE_mapping_inbred.tsv"
[1] "Number of markers with a peak ID"
[1] 1
[1] "I:4884561"
[1] "Number of HD regions that overlap with a QTL region"
[1] 2
[1] "length_Zinc_dichloride"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in file(file, ifelse(append, "a", "w")): cannot open file
'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Zinc_dichloride_HD_region_overlap.tsv':
No such file or directory</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in file(file, ifelse(append, "a", "w")): cannot open the connection&gt;</code></pre>
</div>
</div></div>
</section>
     </main>
<!-- /main column -->  <script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>  </div> <!-- /content --> 
  
</body></html>