{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis 11: Calculate Region HDR Overlaps\n",
    "\n",
    "Based on `DivergentRegion_Overlaps.R` script from Analysis GitHub repo.\n",
    "\n",
    "Take the HVR regions present in the mapping panel and measure the overlap of intervals detected for each toxicant."
   ],
   "id": "00937a5d-bbf9-47dc-8480-7050ae257fa8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.4     ✔ readr     2.1.5\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.1\n",
      "✔ ggplot2   3.5.1     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.3     ✔ tidyr     1.3.1\n",
      "✔ purrr     1.0.2     \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::filter() masks stats::filter()\n",
      "✖ dplyr::lag()    masks stats::lag()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "The following object is masked from 'package:lubridate':\n",
      "\n",
      "    %within%\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "Loading required package: GenomeInfoDb"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    hour, isoweek, mday, minute, month, quarter, second, wday, week,\n",
      "    yday, year\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    transpose"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "------------------------------------------------------------------------------\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "Attaching package: 'plyr'\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    desc\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    rename\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n"
   ],
   "id": "67edaee5-1441-496d-a644-837928bf55a0"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs\n",
    "\n",
    "-   `hd_regions_file`: path to the `hvr_gwas_strains.tsv` file generated by `analysis/hdr_summarize_strain.qmd`\n",
    "-   `ns_folder`: path to the nemascan analysis folder"
   ],
   "id": "ccf48375-ead1-4cf5-8a0f-4a2dfafece7b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# path to HDR regions file for GWAS strains from analysis/hdr_summarize_strain.qmd\n",
    "hd_regions_file <- \"data/processed/HDR_overlap/hvr_gwas_strains.tsv\"\n",
    "\n",
    "# Path to folder with NemaScan results\n",
    "ns_folder <- \"data/processed/20231116_Analysis_NemaScan\"\n"
   ],
   "id": "321a28b9-ae39-4eff-8698-c5ef39362e8b"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputs\n",
    "\n",
    "In the output directory (`data/processed/HVR_Overlap/All_QTL_HD_overlap`) there is a file generated for each trait (e.g `length_2_4_D_HD_region_overlap.tsv`). Each row in the file represented a qtl detected for that trait. These files are used later in the `triage` scripts."
   ],
   "id": "d6e50b48-a232-430c-a993-9b29648cb45f"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in dir.create(outdir): cannot create dir\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap', reason 'No such file or\n",
      "directory'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in dir.create(inbred_out_dir): cannot create dir\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred', reason 'No such file or\n",
      "directory'"
     ]
    }
   ],
   "source": [
    "outdir <- glue::glue(\"data/processed/HVR_Overlap/All_QTL_HD_overlap\")\n",
    "\n",
    "#check if the outdir exists, if not create it\n",
    "if (!dir.exists(outdir)) {\n",
    "    dir.create(outdir)\n",
    "}\n"
   ],
   "id": "467faa9e-4cb4-46e0-b468-5d1f40a82501"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ],
   "id": "52e6d7ff-2f90-4ceb-9a86-94df0ba8bcf2"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "create_proc_mapping <- function(mapping_file) {\n",
    "    mapping <- fread(mapping_file)\n",
    "    mapping <- mapping %>%\n",
    "        filter(!is.na(peak_id))\n",
    "    \n",
    "    n_peak_markers <- length(unique(mapping$peak_id))\n",
    "    print(\"Number of markers with a peak ID\")\n",
    "    print(n_peak_markers)\n",
    "    \n",
    "    if (n_peak_markers == 0) {\n",
    "        mapping_list <- NULL\n",
    "    }\n",
    "    mapping_list <- list()\n",
    "    for (i in 1:n_peak_markers) {\n",
    "        \n",
    "        \n",
    "        fil_mapping <- mapping %>%\n",
    "            filter(peak_id == i)\n",
    "        \n",
    "        peak_marker <- unique(fil_mapping$marker)\n",
    "\n",
    "        ref_strains <- fil_mapping %>%\n",
    "            filter(allele == -1) %>%\n",
    "            distinct(strain) %>%\n",
    "            pull(strain)\n",
    "        \n",
    "        alt_strains <- fil_mapping %>%\n",
    "            filter(allele == 1) %>%\n",
    "            distinct(strain) %>%\n",
    "            pull(strain)\n",
    "        \n",
    "        single_peak_df <- fil_mapping  %>% \n",
    "            select(CHROM, startPOS, endPOS) %>%\n",
    "            distinct()\n",
    "        \n",
    "        #Create a genomic ranges object for the peak\n",
    "        peak_gr <- GRanges(seqnames = single_peak_df$CHROM,\n",
    "                           ranges = IRanges(start = single_peak_df$startPOS,\n",
    "                                            end = single_peak_df$endPOS),\n",
    "\n",
    "        )\n",
    "\n",
    "        peak_data <- list(ref_strains = ref_strains, alt_strains = alt_strains, peak_gr = peak_gr)\n",
    "\n",
    "        mapping_list[[peak_marker]] <- peak_data\n",
    "    }\n",
    "    \n",
    "    return(mapping_list)\n",
    "}\n",
    "\n",
    "# Function to create a genomic ranges object for the HD regions for the strain set\n",
    "\n",
    "create_hd_gr <- function(hd_file) {\n",
    "    hd_df <- data.table::fread(\n",
    "        hd_file,\n",
    "        col.names = c(\"chrom\", \"start\", \"stop\", \"hd_region\", \"strains\", \"n_strains\")\n",
    "    )\n",
    "    hd_gr <- GRanges(seqnames = hd_df$chrom,\n",
    "                     ranges = IRanges(start = hd_df$start,\n",
    "                                      end = hd_df$end),\n",
    "                     hd_region = hd_df$hd_region,\n",
    "                     strains = hd_df$strains,\n",
    "                     n_strains = hd_df$n_strains\n",
    "                     )\n",
    "    return(hd_gr)\n",
    "}\n",
    "\n",
    "#Create a function to check if a peak overlaps with a HD region\n",
    "strains_with_hd_overlap <- function(peak_gr, hd_gr) {\n",
    "    #return only the ranges in the hd_gr that overlap with the peak_gr\n",
    "    hd_overlap <- subsetByOverlaps(hd_gr, peak_gr)\n",
    "    \n",
    "    n_hd_overlaps <- length(hd_overlap)\n",
    "    print( \"Number of HD regions that overlap with a QTL region\")\n",
    "    print(n_hd_overlaps)\n",
    "\n",
    "    #return the hd_overlap object as a data frame\n",
    "    hd_overlap_df <- as.data.frame(hd_overlap) %>% \n",
    "    #split the strains column into a list\n",
    "        separate_rows(strains, sep = \",\") \n",
    "\n",
    "    hd_overlap_strains <- hd_overlap_df %>%\n",
    "        select(strains) %>%\n",
    "        distinct() %>% \n",
    "        pull(strains)\n",
    "    \n",
    "    return(hd_overlap_strains)\n",
    "}\n",
    "\n",
    "#Load in the mapping data\n",
    "#example_mapping_file <- \"/projects/b1059/projects/Tim/2021_GWA_manuscript/data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carboxin_AGGREGATE_mapping_inbred.tsv\" \n",
    "#mapping_list <- create_proc_mapping(example_mapping_file)\n",
    "#head(mapping_list)\n",
    "\n",
    "#Load in the HD regions data\n",
    "#hd_regions_file <- \"data/20231121_divergent_regions_strain.bed\"\n",
    "#hd_gr <- create_hd_gr(hd_regions_file)\n",
    "#head(hd_gr)\n",
    "\n",
    "create_trait_overlaps <- function(mapping_file, hd_regions_file, outdir){\n",
    "\n",
    "#Create mapping files obj\n",
    "mapping_list <- create_proc_mapping(mapping_file)\n",
    "#if mapping list is null, return null\n",
    "if (is.null(mapping_list)) {\n",
    "    stop(\"No markers with peak IDs\")\n",
    "}\n",
    "\n",
    "#Create HD regions obj\n",
    "hd_gr <- create_hd_gr(hd_regions_file)\n",
    "\n",
    "#Iterate through the mapping list and check for HD region overlaps\n",
    "hd_overlap_list <- list()\n",
    "\n",
    "for (i in 1:length(mapping_list)) {\n",
    "    peak_data <- mapping_list[[i]]\n",
    "    #get the marker id for the peak from the name of the list element\n",
    "    peak_marker <- names(mapping_list[i])\n",
    "    print(peak_marker)\n",
    "    peak_gr <- peak_data$peak_gr\n",
    "    \n",
    "    #calculate the number of strains with either ref or alt status at the peak marker\n",
    "    n_strains_ref <- length(peak_data$ref_strains)\n",
    "    n_strains_alt <- length(peak_data$alt_strains)\n",
    "\n",
    "    total_strains <- n_strains_ref + n_strains_alt\n",
    "    \n",
    "    hd_overlap_strains <- strains_with_hd_overlap(peak_gr, hd_gr)\n",
    "    \n",
    "    n_strains_ref_overlap <- length(intersect(peak_data$ref_strains, hd_overlap_strains))\n",
    "    n_strains_alt_overlap <- length(intersect(peak_data$alt_strains, hd_overlap_strains))\n",
    "\n",
    "    hd_overlap_list[[peak_marker]] <- list(\n",
    "                                            peak_marker = peak_marker,\n",
    "                                            n_strains_ref = n_strains_ref,\n",
    "                                            n_strains_alt = n_strains_alt,\n",
    "                                            total_strains = total_strains,\n",
    "                                            n_strains_ref_overlap = n_strains_ref_overlap,\n",
    "                                            n_strains_alt_overlap = n_strains_alt_overlap,\n",
    "                                            prop_ref_overlap = n_strains_ref_overlap/n_strains_ref,\n",
    "                                            prop_alt_overlap = n_strains_alt_overlap/n_strains_alt\n",
    "                                           )\n",
    "}\n",
    "\n",
    "head(hd_overlap_list)\n",
    "\n",
    "hd_overlap_df <- ldply(hd_overlap_list, data.frame)\n",
    "\n",
    "#extract the trait name from the mapping file name\n",
    "get_traitname <- function(fn){\n",
    "    bn <- basename(fn)\n",
    "    if(str_detect(bn, \"CV_length\")){\n",
    "        #extract everthing before _AGGREGATE\n",
    "        traitname <- str_extract(bn, \".*(?=_AGGREGATE)\")\n",
    "        traitname <- str_remove(traitname, \"processed_\")\n",
    "    }else{\n",
    "        #extract everything before _AGGREGATE \n",
    "        traitname <- str_extract(bn, \".*(?=_AGGREGATE)\")\n",
    "        traitname <- str_remove(traitname, \"processed_\")\n",
    "    }\n",
    "    return(traitname)\n",
    "}\n",
    "trait_name <- get_traitname(mapping_file)\n",
    "\n",
    "print(trait_name)\n",
    "\n",
    "hd_overlap_df$trait_name <- trait_name\n",
    "\n",
    "#Add the algoirthm to the data frame\n",
    "get_algorithm <- function(fn){\n",
    "    bn <- basename(fn)\n",
    "    if(str_detect(bn, \"inbred\")){\n",
    "        algorithm <- \"inbred\"\n",
    "    }else{\n",
    "        algorithm <- \"loco\"\n",
    "    }\n",
    "    return(algorithm)\n",
    "}\n",
    "algorithm <- get_algorithm(mapping_file)\n",
    "hd_overlap_df$algorithm <- algorithm\n",
    "\n",
    "\n",
    "head(hd_overlap_df)\n",
    "\n",
    "#write out the data frame with the HD region overlap data using the trait name\n",
    "#hd_overlap_outfile <- paste0(\"data/\", trait_name, \"_HD_region_overlap.tsv\")\n",
    "hd_overlap_outfile <- glue::glue(\"{outdir}/{trait_name}_HD_region_overlap.tsv\")\n",
    "\n",
    "# Check if the file exists\n",
    "if (file.exists(hd_overlap_outfile)) {\n",
    "    # If the file exists, read the existing data\n",
    "    existing_data <- read.table(hd_overlap_outfile, header = TRUE, sep = \"\\t\")\n",
    "    \n",
    "    # Append the new data to the existing data\n",
    "    combined_data <- rbind(existing_data, hd_overlap_df)\n",
    "} else {\n",
    "    # If the file does not exist, the new data is the combined data\n",
    "    combined_data <- hd_overlap_df\n",
    "}\n",
    "\n",
    "# Write the combined data to the file\n",
    "write.table(combined_data, hd_overlap_outfile, sep = \"\\t\", row.names = FALSE)\n",
    "\n",
    "}\n"
   ],
   "id": "c56e71c6-603f-4a1c-8c9f-f0b0028f8f8b"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main\n",
    "\n",
    "The main function executed by this script is `create_trait_overlaps()` which executes a series of other functions defined in the script.\n",
    "\n",
    "1.  Create a mapping files object (a list with several named elements) with `create_proc_mapping()`\n",
    "    -   Function loads the mapping file containing the p-values for all markers and filters to just the peak marker for each interval that was defined by NS.\n",
    "    -   Using the filtered mapping data the function counts a list of strains with the REF or ALT allele peak marker allele. Generates the `ref_strains` and `alt_strains` elements of the mapping files object.\n",
    "    -   Creates a genomic ranges object using data about the interval that peak marker represents. This is the `peak_gr` element of the mapping files object.\n",
    "2.  Create a genomic range object from the HVR bedfile using the `create_hd_gr()` function.\n",
    "3.  Identify strains with that have an HVR in the region of the peak marker using the `strains_with_hd_overlap()` function.\n",
    "4.  Calculate the proportion of strains with each peak marker allele REF or ALT that have an HVR overlap."
   ],
   "id": "a1f0ebfe-3db3-4ccf-8731-6032fc7bbe49"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_2_4_D_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"V:14453636\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 10\n",
      "[1] \"length_2_4_D\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_2_4_D_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Aldicarb_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 5\n",
      "[1] \"III:8311919\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 0\n",
      "[1] \"IV:2368192\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 6\n",
      "[1] \"IV:13208614\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 2\n",
      "[1] \"IV:16631628\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 29\n",
      "[1] \"X:13252751\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 8\n",
      "[1] \"length_Aldicarb\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Aldicarb_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Arsenic_trioxide_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 2\n",
      "[1] \"II:13692527\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 76\n",
      "[1] \"III:1642838\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 198\n",
      "[1] \"length_Arsenic_trioxide\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Arsenic_trioxide_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Atrazine_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"II:12658388\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 36\n",
      "[1] \"length_Atrazine\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Atrazine_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Cadmium_dichloride_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"X:5955567\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 14\n",
      "[1] \"length_Cadmium_dichloride\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Cadmium_dichloride_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carbaryl_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"I:2596756\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 13\n",
      "[1] \"length_Carbaryl\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Carbaryl_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Carboxin_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"III:2527992\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 32\n",
      "[1] \"length_Carboxin\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Carboxin_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorfenapyr_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorothalonil_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 6\n",
      "[1] \"I:1741155\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 13\n",
      "[1] \"II:5392976\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 88\n",
      "[1] \"II:13921653\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 63\n",
      "[1] \"III:1455164\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 51\n",
      "[1] \"III:2605810\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 36\n",
      "[1] \"III:4478684\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 6\n",
      "[1] \"length_Chlorothalonil\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Chlorothalonil_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Chlorpyrifos_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Copper_chloride_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_DMSO_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Lead_nitrate_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Malathion_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Mancozeb_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"II:10305644\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 7\n",
      "[1] \"length_Mancozeb\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Mancozeb_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Methomyl_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Methyl_mercury_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 8\n",
      "[1] \"I:2994073\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 5\n",
      "[1] \"II:626831\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 44\n",
      "[1] \"II:2730209\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 21\n",
      "[1] \"II:3977000\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 32\n",
      "[1] \"III:2461452\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 35\n",
      "[1] \"III:3716201\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 1\n",
      "[1] \"III:10133612\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 7\n",
      "[1] \"V:15476283\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 61\n",
      "[1] \"length_Methyl_mercury\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Methyl_mercury_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Nickel_dichloride_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 2\n",
      "[1] \"II:12203935\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 8\n",
      "[1] \"V:1486141\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 15\n",
      "[1] \"length_Nickel_dichloride\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Nickel_dichloride_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Paraquat_250_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Paraquat_62_5_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 4\n",
      "[1] \"I:3282224\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 0\n",
      "[1] \"I:11258815\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 11\n",
      "[1] \"II:13675568\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 57\n",
      "[1] \"X:2848181\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 9\n",
      "[1] \"length_Paraquat_62_5\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Paraquat_62_5_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Propoxur_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"V:5315004\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 14\n",
      "[1] \"length_Propoxur\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Propoxur_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Pyraclostrobin_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Silver_nitrate_250_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Silver_nitrate_7_8_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 3\n",
      "[1] \"I:1622740\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 17\n",
      "[1] \"II:3016298\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 128\n",
      "[1] \"II:11883855\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 6\n",
      "[1] \"length_Silver_nitrate_7_8\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Silver_nitrate_7_8_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Triphenyl_phosphate_50_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Triphenyl_phosphate_6_25_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 2\n",
      "[1] \"IV:16034033\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 5\n",
      "[1] \"V:7623783\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 229\n",
      "[1] \"length_Triphenyl_phosphate_6_25\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Triphenyl_phosphate_6_25_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Water_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 0\n",
      "<simpleError in mapping_list[[peak_marker]] <- peak_data: attempt to select less than one element in OneIndex>\n",
      "[1] \"data/processed/20231116_Analysis_NemaScan/INBRED/Mapping/Processed/processed_length_Zinc_dichloride_AGGREGATE_mapping_inbred.tsv\"\n",
      "[1] \"Number of markers with a peak ID\"\n",
      "[1] 1\n",
      "[1] \"I:4884561\"\n",
      "[1] \"Number of HD regions that overlap with a QTL region\"\n",
      "[1] 2\n",
      "[1] \"length_Zinc_dichloride\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning in file(file, ifelse(append, \"a\", \"w\")): cannot open file\n",
      "'data/processed/HVR_Overlap/All_QTL_HD_overlap/inbred/length_Zinc_dichloride_HD_region_overlap.tsv':\n",
      "No such file or directory"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "<simpleError in file(file, ifelse(append, \"a\", \"w\")): cannot open the connection>"
     ]
    }
   ],
   "source": [
    "#Get the mapping directories\n",
    "inbred_mapping_dir <- glue::glue(\"{ns_folder}/INBRED/Mapping/Processed\")\n",
    "\n",
    "#Get the mapping files\n",
    "inbred_mapping_files <- list.files(\n",
    "    path = inbred_mapping_dir,\n",
    "    pattern = \"AGGREGATE_mapping_inbred\",\n",
    "    recursive = T, full.names = T)\n",
    "\n",
    "#Get rid of CV length files\n",
    "inbred_mapping_files <- inbred_mapping_files[!grepl(\"_CV_\", inbred_mapping_files)]\n",
    "\n",
    "\n",
    "#Iterate through the mapping files and create the HD region overlap data\n",
    "for (i in 1:length(inbred_mapping_files)) {\n",
    "    mapping_file <- inbred_mapping_files[i]\n",
    "    print(mapping_file)\n",
    "    tryCatch(create_trait_overlaps(mapping_file, hd_regions_file, inbred_out_dir),\n",
    "                error = function(e) print(e)\n",
    "                )\n",
    "    #create_trait_overlaps(mapping_file, hd_regions_file, outdir)\n",
    "}\n"
   ],
   "id": "f830c0f1-d413-47aa-ae94-1de31f3d5524"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
