{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis 18: Semantic Similarity for All Genes"
   ],
   "id": "0c71b2c7-9f5c-4fab-a326-a4d2dcd66c68"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'dplyr'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'data.table'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    between, first, last"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "here() starts at /Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript"
     ]
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "$strain_table\n",
      "$strain_table$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S1/table_S1.html\"\n",
      "\n",
      "$strain_table$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S1/table_S1.tsv.zip\"\n",
      "\n",
      "$strain_table$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S1/table_S1.docx\"\n",
      "\n",
      "$strain_table$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S1/table_S1.csv\"\n",
      "\n",
      "\n",
      "$tox_table_ft\n",
      "$tox_table_ft$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S2/table_S2.html\"\n",
      "\n",
      "$tox_table_ft$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S2/table_S2.tsv.zip\"\n",
      "\n",
      "$tox_table_ft$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S2/table_S2.docx\"\n",
      "\n",
      "$tox_table_ft$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S2/table_S2.csv\"\n",
      "\n",
      "\n",
      "$trait_table\n",
      "$trait_table$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S3/table_S3.html\"\n",
      "\n",
      "$trait_table$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S3/table_S3.tsv.zip\"\n",
      "\n",
      "$trait_table$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S3/table_S3.docx\"\n",
      "\n",
      "$trait_table$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S3/table_S3.csv\"\n",
      "\n",
      "\n",
      "$outlier_ft\n",
      "$outlier_ft$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S4/table_S4.html\"\n",
      "\n",
      "$outlier_ft$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S4/table_S4.tsv.zip\"\n",
      "\n",
      "$outlier_ft$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S4/table_S4.docx\"\n",
      "\n",
      "$outlier_ft$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S4/table_S4.csv\"\n",
      "\n",
      "\n",
      "$group_summary\n",
      "$group_summary$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S5/table_S5.html\"\n",
      "\n",
      "$group_summary$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S5/table_S5.tsv.zip\"\n",
      "\n",
      "$group_summary$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S5/table_S5.docx\"\n",
      "\n",
      "$group_summary$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S5/table_S5.csv\"\n",
      "\n",
      "\n",
      "$heritability_summary\n",
      "$heritability_summary$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S6/table_S6.html\"\n",
      "\n",
      "$heritability_summary$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S6/table_S6.tsv.zip\"\n",
      "\n",
      "$heritability_summary$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S6/table_S6.docx\"\n",
      "\n",
      "$heritability_summary$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S6/table_S6.csv\"\n",
      "\n",
      "\n",
      "$qtl_detected_table\n",
      "$qtl_detected_table$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S7/table_S7.html\"\n",
      "\n",
      "$qtl_detected_table$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S7/table_S7.tsv.zip\"\n",
      "\n",
      "$qtl_detected_table$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S7/table_S7.docx\"\n",
      "\n",
      "$qtl_detected_table$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S7/table_S7.csv\"\n",
      "\n",
      "\n",
      "$candidate_genes\n",
      "$candidate_genes$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S8/table_S8.html\"\n",
      "\n",
      "$candidate_genes$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S8/table_S8.tsv.zip\"\n",
      "\n",
      "$candidate_genes$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S8/table_S8.docx\"\n",
      "\n",
      "$candidate_genes$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S8/table_S8.csv\"\n",
      "\n",
      "\n",
      "$overlap_ft\n",
      "$overlap_ft$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S9/table_S9.html\"\n",
      "\n",
      "$overlap_ft$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S9/table_S9.tsv.zip\"\n",
      "\n",
      "$overlap_ft$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S9/table_S9.docx\"\n",
      "\n",
      "$overlap_ft$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S9/table_S9.csv\"\n",
      "\n",
      "\n",
      "$inbred_qtl_ranking_ft\n",
      "$inbred_qtl_ranking_ft$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S10/table_S10.html\"\n",
      "\n",
      "$inbred_qtl_ranking_ft$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S10/table_S10.tsv.zip\"\n",
      "\n",
      "$inbred_qtl_ranking_ft$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S10/table_S10.docx\"\n",
      "\n",
      "$inbred_qtl_ranking_ft$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S10/table_S10.csv\"\n",
      "\n",
      "\n",
      "$orthologs_ft\n",
      "$orthologs_ft$html\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S11/table_S11.html\"\n",
      "\n",
      "$orthologs_ft$tsv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S11/table_S11.tsv.zip\"\n",
      "\n",
      "$orthologs_ft$docx\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S11/table_S11.docx\"\n",
      "\n",
      "$orthologs_ft$csv\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/tables/table_S11/table_S11.csv\"\n",
      "\n",
      "\n",
      "$heritability_all_classes\n",
      "$heritability_all_classes$png\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_1/figure_1.png\"\n",
      "\n",
      "$heritability_all_classes$eps\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_1/figure_1.eps\"\n",
      "\n",
      "$heritability_all_classes$jpg\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_1/figure_1.jpg\"\n",
      "\n",
      "\n",
      "$phenotype_correlations\n",
      "$phenotype_correlations$png\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_2/figure_2.png\"\n",
      "\n",
      "$phenotype_correlations$eps\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_2/figure_2.eps\"\n",
      "\n",
      "$phenotype_correlations$jpg\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_2/figure_2.jpg\"\n",
      "\n",
      "\n",
      "$manhattan_plots\n",
      "$manhattan_plots$png\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_3/figure_3.png\"\n",
      "\n",
      "$manhattan_plots$eps\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_3/figure_3.eps\"\n",
      "\n",
      "$manhattan_plots$jpg\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_3/figure_3.jpg\"\n",
      "\n",
      "\n",
      "$tox.effect.bubble\n",
      "$tox.effect.bubble$png\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_4/figure_4.png\"\n",
      "\n",
      "$tox.effect.bubble$eps\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_4/figure_4.eps\"\n",
      "\n",
      "$tox.effect.bubble$jpg\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_4/figure_4.jpg\"\n",
      "\n",
      "\n",
      "$qtl_overlaps\n",
      "$qtl_overlaps$png\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_5/figure_5.png\"\n",
      "\n",
      "$qtl_overlaps$eps\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_5/figure_5.eps\"\n",
      "\n",
      "$qtl_overlaps$jpg\n",
      "[1] \"/Users/ryanmckeown/Desktop/ToxinGWAS_Manuscript/figures/figure_5/figure_5.jpg\""
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'kableExtra'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    group_rows"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attaching package: 'flextable'"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The following objects are masked from 'package:kableExtra':\n",
      "\n",
      "    as_image, footnote"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ forcats   1.0.0     ✔ readr     2.1.5\n",
      "✔ ggplot2   3.5.1     ✔ stringr   1.5.1\n",
      "✔ lubridate 1.9.3     ✔ tibble    3.2.1\n",
      "✔ purrr     1.0.2     ✔ tidyr     1.3.1"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ data.table::between()    masks dplyr::between()\n",
      "✖ purrr::compose()         masks flextable::compose()\n",
      "✖ dplyr::filter()          masks stats::filter()\n",
      "✖ data.table::first()      masks dplyr::first()\n",
      "✖ kableExtra::group_rows() masks dplyr::group_rows()\n",
      "✖ lubridate::hour()        masks data.table::hour()\n",
      "✖ lubridate::isoweek()     masks data.table::isoweek()\n",
      "✖ dplyr::lag()             masks stats::lag()\n",
      "✖ data.table::last()       masks dplyr::last()\n",
      "✖ lubridate::mday()        masks data.table::mday()\n",
      "✖ lubridate::minute()      masks data.table::minute()\n",
      "✖ lubridate::month()       masks data.table::month()\n",
      "✖ lubridate::quarter()     masks data.table::quarter()\n",
      "✖ lubridate::second()      masks data.table::second()\n",
      "✖ purrr::transpose()       masks data.table::transpose()\n",
      "✖ lubridate::wday()        masks data.table::wday()\n",
      "✖ lubridate::week()        masks data.table::week()\n",
      "✖ lubridate::yday()        masks data.table::yday()\n",
      "✖ lubridate::year()        masks data.table::year()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n",
      "\n",
      "\n",
      "Registered S3 methods overwritten by 'treeio':\n",
      "  method              from    \n",
      "  MRCA.phylo          tidytree\n",
      "  MRCA.treedata       tidytree\n",
      "  Nnode.treedata      tidytree\n",
      "  Ntip.treedata       tidytree\n",
      "  ancestor.phylo      tidytree\n",
      "  ancestor.treedata   tidytree\n",
      "  child.phylo         tidytree\n",
      "  child.treedata      tidytree\n",
      "  full_join.phylo     tidytree\n",
      "  full_join.treedata  tidytree\n",
      "  groupClade.phylo    tidytree\n",
      "  groupClade.treedata tidytree\n",
      "  groupOTU.phylo      tidytree\n",
      "  groupOTU.treedata   tidytree\n",
      "  is.rooted.treedata  tidytree\n",
      "  nodeid.phylo        tidytree\n",
      "  nodeid.treedata     tidytree\n",
      "  nodelab.phylo       tidytree\n",
      "  nodelab.treedata    tidytree\n",
      "  offspring.phylo     tidytree\n",
      "  offspring.treedata  tidytree\n",
      "  parent.phylo        tidytree\n",
      "  parent.treedata     tidytree\n",
      "  root.treedata       tidytree\n",
      "  rootnode.phylo      tidytree\n",
      "  sibling.phylo       tidytree\n",
      "\n",
      "clusterProfiler v4.6.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: 'clusterProfiler'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:flextable':\n",
      "\n",
      "    width\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:lubridate':\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from 'package:lubridate':\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from 'package:glue':\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from 'package:data.table':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'AnnotationDbi'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "GOSemSim v2.24.0  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use GOSemSim in published research, please cite:\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu. Gene Ontology Semantic Similarity Analysis Using GOSemSim. In: Kidder B. (eds) Stem Cell Transcriptional Networks. Methods in Molecular Biology, 2020, 2117:207-215. Humana, New York, NY. doi:10.1007/978-1-0716-0301-7_11\n",
      "\u001b[36m-\u001b[39m Guangchuang Yu, Fei Li, Yide Qin, Xiaochen Bo, Yibo Wu, Shengqi Wang. GOSemSim: an R package for measuring semantic similarity among GO terms and gene products Bioinformatics 2010, 26(7):976-978. doi:10.1093/bioinformatics/btq064"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "preparing gene to GO mapping data...\n",
      "preparing IC data..."
     ]
    }
   ],
   "source": [
    "\n",
    "library(dplyr)\n"
   ],
   "id": "1106d99b-ea50-47a1-a1ea-e3be4a962482"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs"
   ],
   "id": "3e708bfd-03b6-4e20-8222-25e649c4cf0c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to NS output directory\n",
    "ns_dir <- \"data/processed/20231116_Analysis_NemaScan\"\n",
    "tox_file <- \"data/processed/tox_data/tox_metadata.csv\"\n"
   ],
   "id": "3eedffd9-c794-42ee-8237-e7c494c3da88"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputs"
   ],
   "id": "80002084-6c63-410b-b8c4-3b531bf8b96d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Output directory 'data/processed/semantic_similarity' already exists."
     ]
    }
   ],
   "source": [
    "# path to output directory should be the same name as the dir contain the script\n",
    "out_dir <- \"data/processed/semantic_similarity\"\n",
    "out_fn <- \"all_genes_semSim.rds\"\n",
    "# check if the output directory exists, if not create it\n",
    "if (!dir.exists(out_dir)) {\n",
    "  message(glue(\"Output directory '{out_dir}' does not exist. Creating it...\"))\n",
    "  dir.create(out_dir, recursive = TRUE)\n",
    "  message(glue(\"Output directory '{out_dir}' created successfully.\"))\n",
    "} else {\n",
    "  message(glue(\"Output directory '{out_dir}' already exists.\"))\n",
    "}\n"
   ],
   "id": "51931256-208b-4d73-ab71-d5ec4c29cba1"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ],
   "id": "909cf728-99a9-4d5f-b2c1-a78ac85b9f44"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a function to process each trait\n",
    "process_trait <- function(trait_id) {\n",
    "  message(glue(\"Processing trait: {trait_id}\"))\n",
    "  genes <- get_genes_all_intervals(\n",
    "    qtl = inbred,\n",
    "    trait_id = trait_id,\n",
    "    gff_df = gff,\n",
    "    cEGO = cEGO\n",
    "  )\n",
    "  return(genes)\n",
    "}\n"
   ],
   "id": "1396f3e1-e1b7-4da8-93a1-765bbe64197c"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main"
   ],
   "id": "a59fb395-9ce6-4d50-9bf0-6174fcee3640"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Processing 16 traits using 7 cores..."
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "All genes semantic similarity saved to data/processed/semantic_similarity/all_genes_semSim.rds"
     ]
    }
   ],
   "source": [
    "\n",
    "## Load the tox data\n",
    "tox <- data.table::fread(tox_file)\n",
    "\n",
    "# load dataframe with the inbred interval data\n",
    "# [ ] - will need a way to connect the trait/drug to the tox metadata\n",
    "inbred <- data.table::fread(\n",
    "  glue::glue(\"{ns_dir}/INBRED/Mapping/Processed/QTL_peaks_inbred.tsv\"),\n",
    "  data.table = F\n",
    ") %>%\n",
    "  # fix drug names\n",
    "  dplyr::mutate(\n",
    "    drug = stringr::str_replace(\n",
    "      trait,\n",
    "      pattern = \"^length_\",\n",
    "      replacement = \"\"\n",
    "    )\n",
    "  ) %>%\n",
    "  # remove the CV_ length_ traits\n",
    "  dplyr::filter(\n",
    "    !stringr::str_detect(drug, \"CV_\")\n",
    "  ) %>%\n",
    "  dplyr::select(\n",
    "    trait,\n",
    "    drug,\n",
    "    CHROM,\n",
    "    marker,\n",
    "    startPOS,\n",
    "    endPOS\n",
    "  ) %>%\n",
    "  # create an id to be used in list\n",
    "  dplyr::mutate(\n",
    "    qtl.id = paste(trait, marker, sep = \"_\")\n",
    "  )\n",
    "# Extract all unique combinations of the 'trait' column\n",
    "unique_trait_ids <- inbred %>%\n",
    "  dplyr::pull(trait) %>%\n",
    "  unique()\n",
    "\n",
    "# Get number of cores (leave one core free for system)\n",
    "n_cores <- parallel::detectCores() - 1\n",
    "\n",
    "# Process all traits in parallel\n",
    "message(glue(\"Processing {length(unique_trait_ids)} traits using {n_cores} cores...\"))\n"
   ],
   "id": "b0d1cce6-7142-473c-b4e9-f9fb68258d9c"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
